{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extracting and analysing data related to an AOP of interest\n",
    "\n",
    "Citation: Marvin Martens, Thomas Exner, Tomaž Mohorič, Chris T Evelo, Egon L Willighagen. Workflow for extracting and analyzing data related to an AOP of interest. 2020\n",
    "\n",
    "One of the main questions to solve in [AOPLink](https://openrisknet.org/e-infrastructure/development/case-studies/case-study-aoplink/) is the finding of data that supports an AOP of interest. To answer that, we have developed this Jupyter notebook that does that by using a variety of OpenRiskNet services:\n",
    "\n",
    "- AOP-Wiki RDF\n",
    "- ChemIdConvert\n",
    "- AOP-DB RDF\n",
    "- BridgeDb\n",
    "- EdelweissData explorer\n",
    "- WikiPathways\n",
    "\n",
    "After selecting an AOP of interest, information is extracted from the AOP-Wiki RDF, ChemIdConvert, and AOP-DB RDF, to get a better understanding of the AOP. Next, the EdelweissData explorer was used to search for fitting data sets from ToxCast and TG-GATES based on the genes and compounds linked to the AOP of interest. The final part involves pathway analysis using the transcriptomics data of TG-GATES and the molecular pathways of WikiPathways, identifying significantly affected pathways upon exposure to the chemicals of interest."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to execute the Jupyter notebook, a set of Python libraries are required. The following section should import, or install, all of them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting pip\n",
      "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/57/36/67f809c135c17ec9b8276466cc57f35b98c240f55c780689ea29fa32f512/pip-20.0.1-py2.py3-none-any.whl (1.5MB)\n",
      "\u001b[K     |████████████████████████████████| 1.5MB 7.1MB/s \n",
      "\u001b[?25hInstalling collected packages: pip\n",
      "  Found existing installation: pip 19.3.1\n",
      "    Uninstalling pip-19.3.1:\n",
      "      Successfully uninstalled pip-19.3.1\n",
      "Successfully installed pip-20.0.1\n",
      "Collecting watermark\n",
      "  Downloading watermark-2.0.2-py2.py3-none-any.whl (5.3 kB)\n",
      "Requirement already satisfied: ipython in /opt/app-root/lib/python3.6/site-packages (from watermark) (7.9.0)\n",
      "Requirement already satisfied: decorator in /opt/app-root/lib/python3.6/site-packages (from ipython->watermark) (4.4.1)\n",
      "Requirement already satisfied: setuptools>=18.5 in /opt/app-root/lib/python3.6/site-packages (from ipython->watermark) (41.6.0)\n",
      "Requirement already satisfied: pickleshare in /opt/app-root/lib/python3.6/site-packages (from ipython->watermark) (0.7.5)\n",
      "Requirement already satisfied: pexpect; sys_platform != \"win32\" in /opt/app-root/lib/python3.6/site-packages (from ipython->watermark) (4.7.0)\n",
      "Requirement already satisfied: jedi>=0.10 in /opt/app-root/lib/python3.6/site-packages (from ipython->watermark) (0.15.1)\n",
      "Requirement already satisfied: traitlets>=4.2 in /opt/app-root/lib/python3.6/site-packages (from ipython->watermark) (4.3.3)\n",
      "Requirement already satisfied: prompt-toolkit<2.1.0,>=2.0.0 in /opt/app-root/lib/python3.6/site-packages (from ipython->watermark) (2.0.10)\n",
      "Requirement already satisfied: backcall in /opt/app-root/lib/python3.6/site-packages (from ipython->watermark) (0.1.0)\n",
      "Requirement already satisfied: pygments in /opt/app-root/lib/python3.6/site-packages (from ipython->watermark) (2.4.2)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /opt/app-root/lib/python3.6/site-packages (from pexpect; sys_platform != \"win32\"->ipython->watermark) (0.6.0)\n",
      "Requirement already satisfied: parso>=0.5.0 in /opt/app-root/lib/python3.6/site-packages (from jedi>=0.10->ipython->watermark) (0.5.1)\n",
      "Requirement already satisfied: ipython-genutils in /opt/app-root/lib/python3.6/site-packages (from traitlets>=4.2->ipython->watermark) (0.2.0)\n",
      "Requirement already satisfied: six in /opt/app-root/lib/python3.6/site-packages (from traitlets>=4.2->ipython->watermark) (1.12.0)\n",
      "Requirement already satisfied: wcwidth in /opt/app-root/lib/python3.6/site-packages (from prompt-toolkit<2.1.0,>=2.0.0->ipython->watermark) (0.1.7)\n",
      "Installing collected packages: watermark\n",
      "Successfully installed watermark-2.0.2\n",
      "Collecting sparqlwrapper\n",
      "  Downloading SPARQLWrapper-1.8.5-py3-none-any.whl (26 kB)\n",
      "Collecting rdflib>=4.0\n",
      "  Downloading rdflib-4.2.2-py3-none-any.whl (344 kB)\n",
      "\u001b[K     |████████████████████████████████| 344 kB 8.7 MB/s \n",
      "\u001b[?25hCollecting pyparsing\n",
      "  Downloading pyparsing-2.4.6-py2.py3-none-any.whl (67 kB)\n",
      "\u001b[K     |████████████████████████████████| 67 kB 18.1 MB/s \n",
      "\u001b[?25hCollecting isodate\n",
      "  Downloading isodate-0.6.0-py2.py3-none-any.whl (45 kB)\n",
      "\u001b[K     |████████████████████████████████| 45 kB 20.4 MB/s \n",
      "\u001b[?25hRequirement already satisfied: six in /opt/app-root/lib/python3.6/site-packages (from isodate->rdflib>=4.0->sparqlwrapper) (1.12.0)\n",
      "Installing collected packages: pyparsing, isodate, rdflib, sparqlwrapper\n",
      "Successfully installed isodate-0.6.0 pyparsing-2.4.6 rdflib-4.2.2 sparqlwrapper-1.8.5\n",
      "Collecting pyvis\n",
      "  Downloading pyvis-0.1.7.0-py3-none-any.whl (22 kB)\n",
      "Requirement already satisfied: ipython>=5.3.0 in /opt/app-root/lib/python3.6/site-packages (from pyvis) (7.9.0)\n",
      "Requirement already satisfied: jinja2>=2.9.6 in /opt/app-root/lib/python3.6/site-packages (from pyvis) (2.10.3)\n",
      "Collecting networkx>=1.11\n",
      "  Downloading networkx-2.4-py3-none-any.whl (1.6 MB)\n",
      "\u001b[K     |████████████████████████████████| 1.6 MB 8.4 MB/s \n",
      "\u001b[?25hRequirement already satisfied: jedi>=0.10 in /opt/app-root/lib/python3.6/site-packages (from ipython>=5.3.0->pyvis) (0.15.1)\n",
      "Requirement already satisfied: pexpect; sys_platform != \"win32\" in /opt/app-root/lib/python3.6/site-packages (from ipython>=5.3.0->pyvis) (4.7.0)\n",
      "Requirement already satisfied: backcall in /opt/app-root/lib/python3.6/site-packages (from ipython>=5.3.0->pyvis) (0.1.0)\n",
      "Requirement already satisfied: pygments in /opt/app-root/lib/python3.6/site-packages (from ipython>=5.3.0->pyvis) (2.4.2)\n",
      "Requirement already satisfied: decorator in /opt/app-root/lib/python3.6/site-packages (from ipython>=5.3.0->pyvis) (4.4.1)\n",
      "Requirement already satisfied: pickleshare in /opt/app-root/lib/python3.6/site-packages (from ipython>=5.3.0->pyvis) (0.7.5)\n",
      "Requirement already satisfied: prompt-toolkit<2.1.0,>=2.0.0 in /opt/app-root/lib/python3.6/site-packages (from ipython>=5.3.0->pyvis) (2.0.10)\n",
      "Requirement already satisfied: traitlets>=4.2 in /opt/app-root/lib/python3.6/site-packages (from ipython>=5.3.0->pyvis) (4.3.3)\n",
      "Requirement already satisfied: setuptools>=18.5 in /opt/app-root/lib/python3.6/site-packages (from ipython>=5.3.0->pyvis) (41.6.0)\n",
      "Requirement already satisfied: MarkupSafe>=0.23 in /opt/app-root/lib/python3.6/site-packages (from jinja2>=2.9.6->pyvis) (1.1.1)\n",
      "Requirement already satisfied: parso>=0.5.0 in /opt/app-root/lib/python3.6/site-packages (from jedi>=0.10->ipython>=5.3.0->pyvis) (0.5.1)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /opt/app-root/lib/python3.6/site-packages (from pexpect; sys_platform != \"win32\"->ipython>=5.3.0->pyvis) (0.6.0)\n",
      "Requirement already satisfied: wcwidth in /opt/app-root/lib/python3.6/site-packages (from prompt-toolkit<2.1.0,>=2.0.0->ipython>=5.3.0->pyvis) (0.1.7)\n",
      "Requirement already satisfied: six>=1.9.0 in /opt/app-root/lib/python3.6/site-packages (from prompt-toolkit<2.1.0,>=2.0.0->ipython>=5.3.0->pyvis) (1.12.0)\n",
      "Requirement already satisfied: ipython-genutils in /opt/app-root/lib/python3.6/site-packages (from traitlets>=4.2->ipython>=5.3.0->pyvis) (0.2.0)\n",
      "Installing collected packages: networkx, pyvis\n",
      "Successfully installed networkx-2.4 pyvis-0.1.7.0\n",
      "Collecting simplejson\n",
      "  Downloading simplejson-3.17.0.tar.gz (83 kB)\n",
      "\u001b[K     |████████████████████████████████| 83 kB 6.7 MB/s \n",
      "\u001b[?25hBuilding wheels for collected packages: simplejson\n",
      "  Building wheel for simplejson (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for simplejson: filename=simplejson-3.17.0-cp36-cp36m-linux_x86_64.whl size=111916 sha256=9416e23f3d68d8c6a642e5994636621d06b1d445da8b925b0aa32cb163cec392\n",
      "  Stored in directory: /tmp/pip-ephem-wheel-cache-v0ly96hg/wheels/c2/95/a2/477b46cfe980061fddc4cb76ed4657eceec35d2b2090355e41\n",
      "Successfully built simplejson\n",
      "Installing collected packages: simplejson\n",
      "Successfully installed simplejson-3.17.0\n",
      "Collecting pandas\n",
      "  Downloading pandas-0.25.3-cp36-cp36m-manylinux1_x86_64.whl (10.4 MB)\n",
      "\u001b[K     |████████████████████████████████| 10.4 MB 7.5 MB/s \n",
      "\u001b[?25hRequirement already satisfied: python-dateutil>=2.6.1 in /opt/app-root/lib/python3.6/site-packages (from pandas) (2.8.1)\n",
      "Collecting numpy>=1.13.3\n",
      "  Downloading numpy-1.18.1-cp36-cp36m-manylinux1_x86_64.whl (20.1 MB)\n",
      "\u001b[K     |████████████████████████████████| 20.1 MB 24.2 MB/s \n",
      "\u001b[?25hCollecting pytz>=2017.2\n",
      "  Downloading pytz-2019.3-py2.py3-none-any.whl (509 kB)\n",
      "\u001b[K     |████████████████████████████████| 509 kB 24.0 MB/s \n",
      "\u001b[?25hRequirement already satisfied: six>=1.5 in /opt/app-root/lib/python3.6/site-packages (from python-dateutil>=2.6.1->pandas) (1.12.0)\n",
      "Installing collected packages: numpy, pytz, pandas\n",
      "Successfully installed numpy-1.18.1 pandas-0.25.3 pytz-2019.3\n",
      "Collecting edelweiss_data\n",
      "  Downloading edelweiss_data-0.1.5-py3-none-any.whl (10 kB)\n",
      "Requirement already satisfied: requests<3.0.0,>=2.18.4 in /opt/app-root/lib/python3.6/site-packages (from edelweiss_data) (2.22.0)\n",
      "Requirement already satisfied: pandas<1.0,>=0.23.0 in /opt/app-root/lib/python3.6/site-packages (from edelweiss_data) (0.25.3)\n",
      "Collecting pyjwt<2.0.0,>=1.5.3\n",
      "  Downloading PyJWT-1.7.1-py2.py3-none-any.whl (18 kB)\n",
      "Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /opt/app-root/lib/python3.6/site-packages (from requests<3.0.0,>=2.18.4->edelweiss_data) (3.0.4)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /opt/app-root/lib/python3.6/site-packages (from requests<3.0.0,>=2.18.4->edelweiss_data) (1.25.6)\n",
      "Requirement already satisfied: idna<2.9,>=2.5 in /opt/app-root/lib/python3.6/site-packages (from requests<3.0.0,>=2.18.4->edelweiss_data) (2.8)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/app-root/lib/python3.6/site-packages (from requests<3.0.0,>=2.18.4->edelweiss_data) (2019.9.11)\n",
      "Requirement already satisfied: pytz>=2017.2 in /opt/app-root/lib/python3.6/site-packages (from pandas<1.0,>=0.23.0->edelweiss_data) (2019.3)\n",
      "Requirement already satisfied: numpy>=1.13.3 in /opt/app-root/lib/python3.6/site-packages (from pandas<1.0,>=0.23.0->edelweiss_data) (1.18.1)\n",
      "Requirement already satisfied: python-dateutil>=2.6.1 in /opt/app-root/lib/python3.6/site-packages (from pandas<1.0,>=0.23.0->edelweiss_data) (2.8.1)\n",
      "Requirement already satisfied: six>=1.5 in /opt/app-root/lib/python3.6/site-packages (from python-dateutil>=2.6.1->pandas<1.0,>=0.23.0->edelweiss_data) (1.12.0)\n",
      "Installing collected packages: pyjwt, edelweiss-data\n",
      "Successfully installed edelweiss-data-0.1.5 pyjwt-1.7.1\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "\n",
    "!{sys.executable} -m pip install --upgrade pip \n",
    "!{sys.executable} -m pip install watermark\n",
    "\n",
    "try:\n",
    "    from SPARQLWrapper import SPARQLWrapper, JSON\n",
    "except ImportError:\n",
    "    !{sys.executable} -m pip install sparqlwrapper\n",
    "    from SPARQLWrapper import SPARQLWrapper, JSON\n",
    "    \n",
    "try:\n",
    "    from pyvis.network import Network\n",
    "except ImportError:\n",
    "    !{sys.executable} -m pip install pyvis\n",
    "    from pyvis.network import Network\n",
    "\n",
    "try:\n",
    "    from IPython.display import display, HTML, IFrame\n",
    "except ImportError:\n",
    "    !{sys.executable} -m pip install ipython\n",
    "    from IPython.display import display, HTML, IFrame\n",
    "\n",
    "try:\n",
    "    import urllib\n",
    "except ImportError:\n",
    "    !{sys.executable} -m pip install urllib\n",
    "    import urllib\n",
    "\n",
    "try:\n",
    "    import simplejson as json\n",
    "except ImportError:\n",
    "    !{sys.executable} -m pip install simplejson\n",
    "    import simplejson as json\n",
    "\n",
    "try:\n",
    "    import pandas as pd\n",
    "except ImportError:\n",
    "    !{sys.executable} -m pip install pandas\n",
    "    import pandas as pd\n",
    "    \n",
    "try:\n",
    "    import re\n",
    "except ImportError:\n",
    "    !{sys.executable} -m pip install re\n",
    "    import re\n",
    "\n",
    "try:\n",
    "    import requests\n",
    "except ImportError:\n",
    "    !{sys.executable} -m pip install requests\n",
    "    import requests\n",
    "    \n",
    "try:\n",
    "    import warnings\n",
    "except ImportError:\n",
    "    !{sys.executable} -m pip install warnings\n",
    "    import warnings\n",
    "\n",
    "try:\n",
    "    import statistics\n",
    "except ImportError:\n",
    "    !{sys.executable} -m pip install statistics\n",
    "    import statistics\n",
    "\n",
    "try:\n",
    "    from edelweiss_data import API, QueryExpression as Q\n",
    "except ImportError:\n",
    "    !{sys.executable} -m pip install edelweiss_data\n",
    "    from edelweiss_data import API, QueryExpression as Q  \n",
    "    \n",
    "pd.set_option('display.max_colwidth', -1)\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Define the AOP of interest\n",
    "\n",
    "This Jupyter notebook focuses on the AOP of interest, which is based on the identifier of the AOP. The identifiers of AOPs can be found on the [AOP-Wiki website](https://aopwiki.org/aops?direction=asc&sort=id \"AOP-Wiki\")."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "AOPid = \"37\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set service URLs\n",
    "\n",
    "The notebook uses a variety of external services. To keep an overview of these, their URLs are defined at the start of the notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SPARQL endpoint URLs\n",
    "aopwikisparql = SPARQLWrapper(\"http://aopwiki-rdf.prod.openrisknet.org/sparql/\")\n",
    "aopdbsparql = SPARQLWrapper(\"http://aopdb-rdf.prod.openrisknet.org/sparql/\")\n",
    "wikipathwayssparql = SPARQLWrapper(\"http://sparql.wikipathways.org/sparql/\")\n",
    "\n",
    "# ChemIdConvert URL\n",
    "chemidconvert = 'https://chemidconvert.cloud.douglasconnect.com/v1/'\n",
    "\n",
    "# BridgeDB base URL\n",
    "bridgedb = 'http://bridgedb.prod.openrisknet.org/'\n",
    "\n",
    "# EdelweissData API URL\n",
    "edelweiss_api_url = 'https://api.staging.kit.cloud.douglasconnect.com'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# AOP-Wiki RDF\n",
    "\n",
    "## Service description\n",
    "The AOP-Wiki repository is part of the AOP Knowledge Base (AOP-KB), a joint effort of the US-Environmental Protection Agency and European Commission - Joint Research Centre. It is developed to facilitate collaborative AOP development, storage of AOPs, and therefore allow reusing toxicological knowledge for risk assessors. This Case Study has converted the AOP-Wiki XML data into an RDF schema, which has been exposed in a public SPARQL endpoint in the OpenRiskNet e-infrastructure.\n",
    "\n",
    "## Implementation\n",
    "First, general information of the AOP is fetched using a variety of SPARQL queries, using predicates from the [AOP-Wiki RDF schema](https://doi.org/10.6084/m9.figshare.11323685.v1 \"AOP-Wiki\"). \n",
    "This is used for:\n",
    "- Creating an overview table of the AOP of interest\n",
    "- Extending the AOP network with connected AOPs\n",
    "\n",
    "Second, stressor chemicals are retrieved and stored for further analysis and fetching of data."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creating overview table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Properties</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dc:title</th>\n",
       "      <td>PPARalpha-dependent liver cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>foaf:page</th>\n",
       "      <td>http://identifiers.org/aop/37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dc:creator</th>\n",
       "      <td>J. Christopher Corton, Cancer AOP Workgroup. National Health and Environmental Effects Research Laboratory, Office of Research and Development, Integrated Systems Toxicology Division, US Environmental Protection Agency, Research Triangle Park, NC. Corresponding author for wiki entry (corton.chris@epa.gov)\\n</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dcterms:abstract</th>\n",
       "      <td>Several therapeutic agents and industrial chemicals induce liver tumors in rats and mice through the activation of the peroxisome proliferator-activated receptor alpha (PPARa). The molecular and cellular events by which PPARa activators induce rodent hepatocarcinogenesis have been extensively studied and elucidated. The weight of evidence relevant to the hypothesized AOP for PPARa activator-induced rodent hepatocarcinogenesis is summarized here. Chemical-specific and mechanistic data support concordance of temporal and dose&amp;ndash;response relationships for the key events associated with many PPARa activators including a phthalate ester plasticizer di(2-ethylhexyl)phthalate (DEHP) and the drug gemfibrozil. The key events (KE) identified are KE1&amp;ndash;PPARa activation, KE2&amp;ndash;alteration in cell growth pathways, KE3&amp;ndash;perturbation of cell growth and survival including increases in cell proliferation and effects on apoptosis, KE4&amp;ndash;selective clonal expansion of preneoplastic foci cells, which lead to the Adverse Outcome&amp;ndash;increases in hepatocellular adenomas and carcinomas. A number of molecular events were identified which were initially evaluated as possible key events. However, the data were not convincing for these to be key events in the MOA. Rather, although not causal these modulating factors were considered to have the potential to alter the ability of PPARa activators to increase liver cancer. These modulating events include increases in oxidative stress, activation of NF-kB, and inhibition of gap junction intercellular communication. While biologically plausible in humans, the hypothesized key events in the rodent MOA, for PPARa activators, are unlikely to induce liver tumors in humans because of biological differences in responses. This conclusion is based on minimal or no effects observed on growth pathways, hepatocellular proliferation and liver tumors in humans and/or species (including hamsters, guinea pigs and cynomolgous monkeys) that are more appropriate human surrogates than mice and rats at overlapping dose levels.\\n</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aopo:has_key_event</th>\n",
       "      <td>http://identifiers.org/aop.events/1170, http://identifiers.org/aop.events/1171, http://identifiers.org/aop.events/227, http://identifiers.org/aop.events/716, http://identifiers.org/aop.events/719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aopo:has_molecular_initiating_event</th>\n",
       "      <td>http://identifiers.org/aop.events/227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aopo:has_adverse_outcome</th>\n",
       "      <td>http://identifiers.org/aop.events/719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aopo:has_key_event_relationship</th>\n",
       "      <td>http://identifiers.org/aop.relationships/1229, http://identifiers.org/aop.relationships/1230, http://identifiers.org/aop.relationships/1232, http://identifiers.org/aop.relationships/1239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ncit:C54571</th>\n",
       "      <td>http://identifiers.org/aop.stressor/11, http://identifiers.org/aop.stressor/175, http://identifiers.org/aop.stressor/191, http://identifiers.org/aop.stressor/205, http://identifiers.org/aop.stressor/206, http://identifiers.org/aop.stressor/207, http://identifiers.org/aop.stressor/208, http://identifiers.org/aop.stressor/210, http://identifiers.org/aop.stressor/211</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              Properties\n",
       "dc:title                             PPARalpha-dependent liver cancer                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   \n",
       "foaf:page                            http://identifiers.org/aop/37                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      \n",
       "dc:creator                           J. Christopher Corton, Cancer AOP Workgroup. National Health and Environmental Effects Research Laboratory, Office of Research and Development, Integrated Systems Toxicology Division, US Environmental Protection Agency, Research Triangle Park, NC. Corresponding author for wiki entry (corton.chris@epa.gov)\\n                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \n",
       "dcterms:abstract                     Several therapeutic agents and industrial chemicals induce liver tumors in rats and mice through the activation of the peroxisome proliferator-activated receptor alpha (PPARa). The molecular and cellular events by which PPARa activators induce rodent hepatocarcinogenesis have been extensively studied and elucidated. The weight of evidence relevant to the hypothesized AOP for PPARa activator-induced rodent hepatocarcinogenesis is summarized here. Chemical-specific and mechanistic data support concordance of temporal and dose&ndash;response relationships for the key events associated with many PPARa activators including a phthalate ester plasticizer di(2-ethylhexyl)phthalate (DEHP) and the drug gemfibrozil. The key events (KE) identified are KE1&ndash;PPARa activation, KE2&ndash;alteration in cell growth pathways, KE3&ndash;perturbation of cell growth and survival including increases in cell proliferation and effects on apoptosis, KE4&ndash;selective clonal expansion of preneoplastic foci cells, which lead to the Adverse Outcome&ndash;increases in hepatocellular adenomas and carcinomas. A number of molecular events were identified which were initially evaluated as possible key events. However, the data were not convincing for these to be key events in the MOA. Rather, although not causal these modulating factors were considered to have the potential to alter the ability of PPARa activators to increase liver cancer. These modulating events include increases in oxidative stress, activation of NF-kB, and inhibition of gap junction intercellular communication. While biologically plausible in humans, the hypothesized key events in the rodent MOA, for PPARa activators, are unlikely to induce liver tumors in humans because of biological differences in responses. This conclusion is based on minimal or no effects observed on growth pathways, hepatocellular proliferation and liver tumors in humans and/or species (including hamsters, guinea pigs and cynomolgous monkeys) that are more appropriate human surrogates than mice and rats at overlapping dose levels.\\n\n",
       "aopo:has_key_event                   http://identifiers.org/aop.events/1170, http://identifiers.org/aop.events/1171, http://identifiers.org/aop.events/227, http://identifiers.org/aop.events/716, http://identifiers.org/aop.events/719                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                \n",
       "aopo:has_molecular_initiating_event  http://identifiers.org/aop.events/227                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              \n",
       "aopo:has_adverse_outcome             http://identifiers.org/aop.events/719                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              \n",
       "aopo:has_key_event_relationship      http://identifiers.org/aop.relationships/1229, http://identifiers.org/aop.relationships/1230, http://identifiers.org/aop.relationships/1232, http://identifiers.org/aop.relationships/1239                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         \n",
       "ncit:C54571                          http://identifiers.org/aop.stressor/11, http://identifiers.org/aop.stressor/175, http://identifiers.org/aop.stressor/191, http://identifiers.org/aop.stressor/205, http://identifiers.org/aop.stressor/206, http://identifiers.org/aop.stressor/207, http://identifiers.org/aop.stressor/208, http://identifiers.org/aop.stressor/210, http://identifiers.org/aop.stressor/211                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Define all variables as ontology terms present in AOP-Wiki RDF\n",
    "title = 'dc:title'\n",
    "webpage = 'foaf:page'\n",
    "creator = 'dc:creator'\n",
    "abstract = 'dcterms:abstract'\n",
    "key_event = 'aopo:has_key_event'\n",
    "molecular_initiating_event = 'aopo:has_molecular_initiating_event'\n",
    "adverse_outcome = 'aopo:has_adverse_outcome'\n",
    "key_event_relationship = 'aopo:has_key_event_relationship'\n",
    "stressor = 'ncit:C54571'\n",
    "\n",
    "#Create the list of all terms of interest\n",
    "listofterms = [title,webpage,creator,abstract,key_event,molecular_initiating_event,adverse_outcome,key_event_relationship,stressor]\n",
    "\n",
    "#Initiate the DataFrame\n",
    "AOPinfo = pd.DataFrame(columns=['Properties'], index = [list(listofterms)])\n",
    "\n",
    "#Query all terms of interest in the selected AOP\n",
    "for term in listofterms:\n",
    "    sparqlquery = '''\n",
    "    PREFIX ncit: <http://ncicb.nci.nih.gov/xml/owl/EVS/Thesaurus.owl#>\n",
    "    SELECT (group_concat(distinct ?item;separator=\";\") as ?items)\n",
    "    WHERE{\n",
    "    ?AOP_URI a aopo:AdverseOutcomePathway;'''+term+''' ?item.\n",
    "    FILTER (?AOP_URI = aop:'''+AOPid +''')}\n",
    "    '''\n",
    "    aopwikisparql.setQuery(sparqlquery)\n",
    "    aopwikisparql.setReturnFormat(JSON)  \n",
    "    results = aopwikisparql.query().convert()\n",
    "    for result in results[\"results\"][\"bindings\"]:\n",
    "        if 'identifiers.org' in result[\"items\"][\"value\"]:\n",
    "            AOPinfo.at[term,'Properties'] = ', '.join(result[\"items\"][\"value\"].split(';'))\n",
    "        else:\n",
    "            AOPinfo.at[term,'Properties'] = result[\"items\"][\"value\"]\n",
    "display(AOPinfo)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generating AOP network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"700\"\n",
       "            height=\"600\"\n",
       "            src=\"./mygraph.html\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7f334d0b1780>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Generate network from AOP + interlinked AOPs.\n",
    "Key_Events = str(AOPinfo.iat[4,0]).split(', ')\n",
    "\n",
    "#From all the KEs from the selected, get the other AOPs in which they are present, and find all KEs of those other AOPs\n",
    "for Key_Event in Key_Events:\n",
    "    sparqlquery = '''\n",
    "    SELECT ?MIE_ID ?KE_ID ?AO_ID ?KER_ID ?KE_Title\n",
    "    WHERE{\n",
    "    ?KE_URI a aopo:KeyEvent; dcterms:isPartOf ?AOP_URI.\n",
    "    ?AOP_URI aopo:has_key_event ?KE_URI2; aopo:has_molecular_initiating_event ?MIE_URI; aopo:has_adverse_outcome ?AO_URI; aopo:has_key_event_relationship ?KER_URI.\n",
    "    ?KE_URI2 rdfs:label ?KE_ID; dc:title ?KE_Title. \n",
    "    ?MIE_URI rdfs:label ?MIE_ID.\n",
    "    ?AO_URI rdfs:label ?AO_ID.\n",
    "    ?KER_URI rdfs:label ?KER_ID.    \n",
    "    FILTER (?KE_URI = <'''+Key_Event+'''>)}\n",
    "    '''\n",
    "    aopwikisparql.setQuery(sparqlquery)\n",
    "    aopwikisparql.setReturnFormat(JSON)  \n",
    "    results = aopwikisparql.query().convert()\n",
    "\n",
    "    MIEs = set([])\n",
    "    KEs = set([])\n",
    "    KEtitle = {}\n",
    "    AOs = set([])\n",
    "    KERs = set([])\n",
    "    for result in results[\"results\"][\"bindings\"]:\n",
    "        MIEs.add(result[\"MIE_ID\"][\"value\"])\n",
    "        AOs.add(result[\"AO_ID\"][\"value\"])\n",
    "        KEs.add(result[\"KE_ID\"][\"value\"])\n",
    "        KERs.add(result[\"KER_ID\"][\"value\"])\n",
    "        KEtitle[result[\"KE_ID\"][\"value\"]]=result[\"KE_Title\"][\"value\"]\n",
    "\n",
    "#List all intermediate KEs that are not MIEs or AOs\n",
    "KEsIntermediate = []\n",
    "for item in KEs:\n",
    "    if item not in MIEs and item not in AOs:\n",
    "        KEsIntermediate.append(item) \n",
    "\n",
    "#Initiate network figure\n",
    "net= Network(height=\"100%\", width=\"100%\")\n",
    "\n",
    "#Add nodes for Molecular Initiating Events, Key Events, and Adverse Outcomes to the network figure\n",
    "for MIE in MIEs:\n",
    "    net.add_node(MIE, color = 'lightgreen', size = 50, shape = 'circle', font = '20px arial black', title = KEtitle[MIE])\n",
    "for KE in KEsIntermediate:\n",
    "    net.add_node(KE, color = 'khaki', size = 50, shape = 'circle', font = '20px arial black', title = KEtitle[KE])\n",
    "for AO in AOs:\n",
    "    net.add_node(AO, color = 'salmon', size = 50, shape = 'circle', font = '20px arial black', title = KEtitle[AO])\n",
    "\n",
    "#Add all Key Event Relationships to the network figure after querying all KERs for all KEs in AOP-Wiki RDF\n",
    "for KER in KERs:\n",
    "    sparqlquery = '''\n",
    "    SELECT ?KE_UP_ID ?KE_DOWN_ID \n",
    "    WHERE{\n",
    "    ?KER_URI a aopo:KeyEventRelationship; rdfs:label ?KER_ID; aopo:has_upstream_key_event ?KE_UP_URI; aopo:has_downstream_key_event ?KE_DOWN_URI.\n",
    "    ?KE_UP_URI rdfs:label ?KE_UP_ID.\n",
    "    ?KE_DOWN_URI rdfs:label ?KE_DOWN_ID.\n",
    "    FILTER (?KER_ID = \"'''+KER+'''\")}\n",
    "    '''\n",
    "    aopwikisparql.setQuery(sparqlquery)\n",
    "    aopwikisparql.setReturnFormat(JSON)  \n",
    "    results = aopwikisparql.query().convert()\n",
    "    for result in results[\"results\"][\"bindings\"]:\n",
    "        net.add_edge(result[\"KE_UP_ID\"][\"value\"],result[\"KE_DOWN_ID\"][\"value\"],width = 2, color = 'black',label = KER, arrows = 'to')\n",
    "\n",
    "net.show('mygraph.html')\n",
    "IFrame(src='./mygraph.html', width=700, height=600)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Query all chemicals that are part of the selected AOP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chemical_name</th>\n",
       "      <th>CAS_ID</th>\n",
       "      <th>CompTox_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Di(2-ethylhexyl) phthalate</td>\n",
       "      <td>117-81-7</td>\n",
       "      <td>DTXSID5020607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gemfibrozil</td>\n",
       "      <td>25812-30-0</td>\n",
       "      <td>DTXSID0020652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Nafenopin</td>\n",
       "      <td>3771-19-5</td>\n",
       "      <td>DTXSID8020911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bezafibrate</td>\n",
       "      <td>41859-67-0</td>\n",
       "      <td>DTXSID3029869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Fenofibrate</td>\n",
       "      <td>49562-28-9</td>\n",
       "      <td>DTXSID2029874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Pirinixic acid</td>\n",
       "      <td>50892-23-4</td>\n",
       "      <td>DTXSID4020290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Ciprofibrate</td>\n",
       "      <td>52214-84-3</td>\n",
       "      <td>DTXSID8020331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Clofibrate</td>\n",
       "      <td>637-07-0</td>\n",
       "      <td>DTXSID3020336</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Chemical_name      CAS_ID     CompTox_ID\n",
       "0  Di(2-ethylhexyl) phthalate  117-81-7    DTXSID5020607\n",
       "1  Gemfibrozil                 25812-30-0  DTXSID0020652\n",
       "2  Nafenopin                   3771-19-5   DTXSID8020911\n",
       "3  Bezafibrate                 41859-67-0  DTXSID3029869\n",
       "4  Fenofibrate                 49562-28-9  DTXSID2029874\n",
       "5  Pirinixic acid              50892-23-4  DTXSID4020290\n",
       "6  Ciprofibrate                52214-84-3  DTXSID8020331\n",
       "7  Clofibrate                  637-07-0    DTXSID3020336"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sparqlquery = '''\n",
    "PREFIX ncit: <http://ncicb.nci.nih.gov/xml/owl/EVS/Thesaurus.owl#>\n",
    "SELECT ?CAS_ID (fn:substring(?CompTox,33) as ?CompTox_ID) ?Chemical_name\n",
    "WHERE{\n",
    "?AOP_URI a aopo:AdverseOutcomePathway; ncit:C54571 ?Stressor.\n",
    "?Stressor aopo:has_chemical_entity ?Chemical.\n",
    "?Chemical cheminf:CHEMINF_000446 ?CAS_ID; dc:title ?Chemical_name.\n",
    "OPTIONAL {?Chemical cheminf:CHEMINF_000568 ?CompTox.}\n",
    "FILTER (?AOP_URI = aop:'''+AOPid +''')}\n",
    "'''\n",
    "aopwikisparql.setQuery(sparqlquery)\n",
    "aopwikisparql.setReturnFormat(JSON)  \n",
    "results = aopwikisparql.query().convert()\n",
    "\n",
    "Chemical_names = {}\n",
    "CompTox = {}\n",
    "\n",
    "for result in results[\"results\"][\"bindings\"]:\n",
    "    try: CompTox[result[\"CAS_ID\"][\"value\"]] =result[\"CompTox_ID\"][\"value\"]\n",
    "    except: pass\n",
    "for result in results[\"results\"][\"bindings\"]:\n",
    "    try: Chemical_names[result[\"CAS_ID\"][\"value\"]] =result[\"Chemical_name\"][\"value\"]\n",
    "    except: pass\n",
    "\n",
    "Chemdata = pd.DataFrame(columns=['Chemical_name','CAS_ID','CompTox_ID'])\n",
    "for CAS_ID in Chemical_names:\n",
    "    Chemdata = Chemdata.append({\n",
    "        'Chemical_name' : Chemical_names[CAS_ID],\n",
    "        'CAS_ID'        : CAS_ID,\n",
    "        'CompTox_ID'    : CompTox[CAS_ID],\n",
    "        }, ignore_index=True)\n",
    "display(Chemdata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['117-81-7',\n",
       " '25812-30-0',\n",
       " '3771-19-5',\n",
       " '41859-67-0',\n",
       " '49562-28-9',\n",
       " '50892-23-4',\n",
       " '52214-84-3',\n",
       " '637-07-0']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compounds = []\n",
    "for index, row in Chemdata.iterrows():\n",
    "    compounds.append(row['CAS_ID'])\n",
    "compounds"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ChemIdConvert\n",
    "\n",
    "## Service description\n",
    "The ChemIdConverter allows users to submit and translate a variety of chemical descriptors, such as SMILES and InChI, through a REST API.\n",
    "\n",
    "## Implementation\n",
    "Convert selected chemical names and display their chemical structures in a dataframe. It takes CAS IDs as an input, and translates them into Smiles and InChI Keys. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CAS_ID</th>\n",
       "      <th>Image</th>\n",
       "      <th>Smiles</th>\n",
       "      <th>InChIKey</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>117-81-7</td>\n",
       "      <td><img style=\"width:150px\" src=\"https://chemidconvert.cloud.douglasconnect.com/v1/asSvg?smiles=CCCCC%28CC%29COC%28%3DO%29c1ccccc1C%28%3DO%29OCC%28CC%29CCCC\"/></td>\n",
       "      <td>CCCCC(CC)COC(=O)c1ccccc1C(=O)OCC(CC)CCCC</td>\n",
       "      <td>BJQHLKABXJIVAM-UHFFFAOYSA-N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>25812-30-0</td>\n",
       "      <td><img style=\"width:150px\" src=\"https://chemidconvert.cloud.douglasconnect.com/v1/asSvg?smiles=Cc1ccc%28C%29c%28OCCCC%28C%29%28C%29C%28O%29%3DO%29c1\"/></td>\n",
       "      <td>Cc1ccc(C)c(OCCCC(C)(C)C(O)=O)c1</td>\n",
       "      <td>HEMJJKBWTPKOJG-UHFFFAOYSA-N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3771-19-5</td>\n",
       "      <td><img style=\"width:150px\" src=\"https://chemidconvert.cloud.douglasconnect.com/v1/asSvg?smiles=CC%28C%29%28Oc1ccc%28cc1%29C2CCCc3ccccc23%29C%28O%29%3DO\"/></td>\n",
       "      <td>CC(C)(Oc1ccc(cc1)C2CCCc3ccccc23)C(O)=O</td>\n",
       "      <td>XJGBDJOMWKAZJS-UHFFFAOYSA-N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>41859-67-0</td>\n",
       "      <td><img style=\"width:150px\" src=\"https://chemidconvert.cloud.douglasconnect.com/v1/asSvg?smiles=CC%28C%29%28Oc1ccc%28CCNC%28%3DO%29c2ccc%28Cl%29cc2%29cc1%29C%28O%29%3DO\"/></td>\n",
       "      <td>CC(C)(Oc1ccc(CCNC(=O)c2ccc(Cl)cc2)cc1)C(O)=O</td>\n",
       "      <td>IIBYAHWJQTYFKB-UHFFFAOYSA-N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>49562-28-9</td>\n",
       "      <td><img style=\"width:150px\" src=\"https://chemidconvert.cloud.douglasconnect.com/v1/asSvg?smiles=CC%28C%29OC%28%3DO%29C%28C%29%28C%29Oc1ccc%28cc1%29C%28%3DO%29c2ccc%28Cl%29cc2\"/></td>\n",
       "      <td>CC(C)OC(=O)C(C)(C)Oc1ccc(cc1)C(=O)c2ccc(Cl)cc2</td>\n",
       "      <td>YMTINGFKWWXKFG-UHFFFAOYSA-N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>50892-23-4</td>\n",
       "      <td><img style=\"width:150px\" src=\"https://chemidconvert.cloud.douglasconnect.com/v1/asSvg?smiles=Cc1cccc%28Nc2cc%28Cl%29nc%28SCC%28O%29%3DO%29n2%29c1C\"/></td>\n",
       "      <td>Cc1cccc(Nc2cc(Cl)nc(SCC(O)=O)n2)c1C</td>\n",
       "      <td>SZRPDCCEHVWOJX-UHFFFAOYSA-N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>52214-84-3</td>\n",
       "      <td><img style=\"width:150px\" src=\"https://chemidconvert.cloud.douglasconnect.com/v1/asSvg?smiles=CC%28C%29%28Oc1ccc%28cc1%29C2CC2%28Cl%29Cl%29C%28O%29%3DO\"/></td>\n",
       "      <td>CC(C)(Oc1ccc(cc1)C2CC2(Cl)Cl)C(O)=O</td>\n",
       "      <td>KPSRODZRAIWAKH-UHFFFAOYSA-N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>637-07-0</td>\n",
       "      <td><img style=\"width:150px\" src=\"https://chemidconvert.cloud.douglasconnect.com/v1/asSvg?smiles=CCOC%28%3DO%29C%28C%29%28C%29Oc1ccc%28Cl%29cc1\"/></td>\n",
       "      <td>CCOC(=O)C(C)(C)Oc1ccc(Cl)cc1</td>\n",
       "      <td>KNHUKKLJHYUCFP-UHFFFAOYSA-N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compoundstable = pd.DataFrame(columns=['CAS_ID', 'Image', 'Smiles', 'InChIKey'])\n",
    "\n",
    "# Fill \"compounds\" with the \"smiles\" by the compound name.\n",
    "for compound in compounds:\n",
    "    smiles = requests.get(chemidconvert + 'cas/to/smiles', params={'cas': compound}).json()['smiles']\n",
    "    inchikey = requests.get(chemidconvert + 'smiles/to/inchikey', params={'smiles': smiles}).json()['inchikey']\n",
    "    compoundstable = compoundstable.append({'CAS_ID': compound, 'Image': smiles, 'Smiles': smiles, 'InChIKey' : inchikey}, ignore_index=True)\n",
    "\n",
    "def smiles_to_image_html(smiles):  # \"smiles\" shadows \"smiles\" from outer scope, use this function only in \"to_html().\n",
    "    \"\"\"Gets for each smile the image, in HTML.\n",
    "    :param smiles: Takes the “smiles” form “compounds”.\n",
    "    :return: The HTML code for the image of the given smiles.\n",
    "    \"\"\"\n",
    "    return '<img style=\"width:150px\" src=\"' + chemidconvert+'asSvg?smiles='+urllib.parse.quote(smiles)+'\"/>'\n",
    "\n",
    "\n",
    "# Return a HTML table of \"compounds\", after \"compounds\" is fill by \"smiles_to_image_html\".\n",
    "HTML(compoundstable.to_html(escape=False, formatters=dict(Image=smiles_to_image_html)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# AOP-DB RDF\n",
    "\n",
    "## Service description\n",
    "The EPA AOP-DB supports the discovery and development of putative and potential AOPs. Based on public annotations, it integrates AOPs with gene targets, chemicals, diseases, tissues, pathways, species orthology information, ontologies, and gene interactions. The AOP-DB facilitates the translation of AOP biological context, and associates assay, chemical and disease endpoints with AOPs ([Pittman et al., 2018](https://doi.org/10.1016/j.taap.2018.02.006); [Mortensen et al., 2018)](https://doi.org/10.1007/s00335-018-9738-7). The AOP-DB won the first OpenRiskNet implementation challenge of the associated partner program and is therefore integrated into the OpenRiskNet e-infrastructure. After the conversion of the AOP-DB into an RDF schema, its data will be exposed in a Virtuoso SPARQL endpoint.\n",
    "\n",
    "## Implementation\n",
    "Extract all genes related to AOP of interest\n",
    "Find all ToxCast assays linked to those genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['5465', '403654', '19013', '25747']\n"
     ]
    }
   ],
   "source": [
    "Key_Events = str(AOPinfo.iat[4,0]).split(', ')\n",
    "Genes = []\n",
    "#from the KEs, get the AOPs\n",
    "for Key_Event in Key_Events:\n",
    "    sparqlquery = '''\n",
    "    SELECT DISTINCT ?KE_ID ?Entrez_ID WHERE{\n",
    "    ?KE_URI edam:data_1027 ?Entrez_URI. ?Entrez_URI edam:data_1027 ?Entrez_ID.\n",
    "    FILTER (?KE_URI = <'''+Key_Event +'''>)}\n",
    "    '''\n",
    "    aopdbsparql.setQuery(sparqlquery)\n",
    "    aopdbsparql.setReturnFormat(JSON)  \n",
    "    results = aopdbsparql.query().convert()\n",
    "    for result in results[\"results\"][\"bindings\"]:\n",
    "        Genes.append(result[\"Entrez_ID\"][\"value\"])\n",
    "print(Genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BridgeDb to map identifiers\n",
    "\n",
    "## Service description\n",
    "In order to link databases and services that use particular identifiers for genes, proteins, and chemicals, the BridgeDb platform is integrated into the OpenRiskNet e-infrastructure. It allows for identifier mapping between various biological databases for data integration and interoperability ([van Iersel et al., 2010](https://doi.org/10.1186/1471-2105-11-5)).\n",
    "\n",
    "## Implementation\n",
    "The genes from AOP-DB are mapped to identifiers from other databases using BridgeDb. Variable values are filled for ```inputdatasource``` and ```outputdatasource``` identifiers based on BridgeDb's [documentation on system codes](https://www.bridgedb.org/documentation/system-codes/). Also, the species is specified as a value in the variable ```Species```.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entrez</th>\n",
       "      <th>HGNC</th>\n",
       "      <th>Ensembl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5465</td>\n",
       "      <td>PPARA</td>\n",
       "      <td>ENSG00000186951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>403654</td>\n",
       "      <td></td>\n",
       "      <td>ENSCAFG00000000788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19013</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUSG00000022383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>25747</td>\n",
       "      <td></td>\n",
       "      <td>ENSRNOG00000021463</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Entrez   HGNC             Ensembl\n",
       "0  5465    PPARA  ENSG00000186951   \n",
       "1  403654         ENSCAFG00000000788\n",
       "2  19013          ENSMUSG00000022383\n",
       "3  25747          ENSRNOG00000021463"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "inputdatasource = 'L'\n",
    "outputdatasource = ['H','En']\n",
    "Species = ['Human','Dog','Mouse','Rat']\n",
    "Mappings = {}\n",
    "HGNC = []\n",
    "\n",
    "for source in outputdatasource:\n",
    "    Mappings[source] = []\n",
    "    for Entrez in Genes:\n",
    "        for species in Species:\n",
    "            allmappings = re.split('\\t|\\n', requests.get(bridgedb + species + '/xrefs/' + inputdatasource + '/' + Entrez + '?dataSource='+source).text)\n",
    "            if allmappings[0] is not '':\n",
    "                break\n",
    "        Mappings[source].append(allmappings[0])\n",
    "\n",
    "ids = {}\n",
    "for source in Mappings:\n",
    "    ids[source]=[]\n",
    "    for identifier in Mappings[source]:\n",
    "        ids[source].append(identifier)\n",
    "\n",
    "GenesTable = pd.DataFrame(columns=['Entrez','HGNC','Ensembl'])\n",
    "GenesTable['Entrez'] = Genes\n",
    "GenesTable['HGNC'] = ids['H']\n",
    "GenesTable['Ensembl'] = ids['En']\n",
    "\n",
    "display(GenesTable)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# EdelweissData explorer\n",
    "\n",
    "Curated datasets are made available through the EdelweissData Explorer, the main data provisioning tool in the [DataCure case study](https://openrisknet.org/e-infrastructure/development/case-studies/case-study-datacure/) of OpenRiskNet. It is a web-based data explorer tool that gives users the ability to filter, search and extract data through the use of API calls. The EdelweissData Explorer serves data from ToxCast, ToxRefDB, and TG-GATES. \n",
    "\n",
    "Prior to using the EdelweissData explorer, the EdelweissData library is initialized and authenticated.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "api = API(edelweiss_api_url)\n",
    "api.authenticate()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tox21/ToxCast\n",
    "Based on the identified genes and chemicals of interest by the AOP-DB RDF and AOP-Wiki RDF, the EdelweissData library is used to find datasets with those particular target genes and chemicals. \n",
    "\n",
    "First, the assays are searched based on the Entrez gene IDs. These are, along with their metadata, stored in a dataframe.\n",
    "\n",
    "Next, for those assays, all datasets are retrieved that are generated with the compounds related to the AOP."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "      <th>Endpoint name</th>\n",
       "      <th>Biological target</th>\n",
       "      <th>Entrez gene ID for the molecular target</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Gene name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th>version</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>d651ef92-c12e-4eba-8979-8dc3f77fc7f3</th>\n",
       "      <th>1</th>\n",
       "      <td>&lt;PublishedDataset 'd651ef92-c12e-4eba-8979-8dc3f77fc7f3':1 - EPA-InVitroDBV3.2-ATG_PPARa_TRANS_dn_summary_tcpl&gt;</td>\n",
       "      <td>ATG_PPARa_TRANS_dn</td>\n",
       "      <td>regulation of transcription factor activity</td>\n",
       "      <td>5465</td>\n",
       "      <td>PPARA</td>\n",
       "      <td>peroxisome proliferator-activated receptor alpha</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5a9cf864-520d-4ca0-b77f-8333aa8a3d5c</th>\n",
       "      <th>1</th>\n",
       "      <td>&lt;PublishedDataset '5a9cf864-520d-4ca0-b77f-8333aa8a3d5c':1 - EPA-InVitroDBV3.2-ATG_PPRE_CIS_dn_summary_tcpl&gt;</td>\n",
       "      <td>ATG_PPRE_CIS_dn</td>\n",
       "      <td>regulation of transcription factor activity</td>\n",
       "      <td>5465</td>\n",
       "      <td>PPARA</td>\n",
       "      <td>peroxisome proliferator-activated receptor alpha</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cd27c421-8273-41cb-9e11-8c96a2b30e10</th>\n",
       "      <th>1</th>\n",
       "      <td>&lt;PublishedDataset 'cd27c421-8273-41cb-9e11-8c96a2b30e10':1 - EPA-InVitroDBV3.2-NVS_NR_hPPARa_summary_tcpl&gt;</td>\n",
       "      <td>NVS_NR_hPPARa</td>\n",
       "      <td>receptor binding</td>\n",
       "      <td>5465</td>\n",
       "      <td>PPARA</td>\n",
       "      <td>peroxisome proliferator-activated receptor alpha</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291d8013-662b-4b44-aa96-894be4473d59</th>\n",
       "      <th>1</th>\n",
       "      <td>&lt;PublishedDataset '291d8013-662b-4b44-aa96-894be4473d59':1 - EPA-InVitroDBV3.2-ATG_PPRE_CIS_up_summary_tcpl&gt;</td>\n",
       "      <td>ATG_PPRE_CIS_up</td>\n",
       "      <td>regulation of transcription factor activity</td>\n",
       "      <td>5465</td>\n",
       "      <td>PPARA</td>\n",
       "      <td>peroxisome proliferator-activated receptor alpha</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f3ebdf70-976f-4b19-92a1-25b189fa13fa</th>\n",
       "      <th>1</th>\n",
       "      <td>&lt;PublishedDataset 'f3ebdf70-976f-4b19-92a1-25b189fa13fa':1 - EPA-InVitroDBV3.2-ATG_PPARa_TRANS_up_summary_tcpl&gt;</td>\n",
       "      <td>ATG_PPARa_TRANS_up</td>\n",
       "      <td>regulation of transcription factor activity</td>\n",
       "      <td>5465</td>\n",
       "      <td>PPARA</td>\n",
       "      <td>peroxisome proliferator-activated receptor alpha</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                      dataset  \\\n",
       "id                                   version                                                                                                                    \n",
       "d651ef92-c12e-4eba-8979-8dc3f77fc7f3 1        <PublishedDataset 'd651ef92-c12e-4eba-8979-8dc3f77fc7f3':1 - EPA-InVitroDBV3.2-ATG_PPARa_TRANS_dn_summary_tcpl>   \n",
       "5a9cf864-520d-4ca0-b77f-8333aa8a3d5c 1        <PublishedDataset '5a9cf864-520d-4ca0-b77f-8333aa8a3d5c':1 - EPA-InVitroDBV3.2-ATG_PPRE_CIS_dn_summary_tcpl>      \n",
       "cd27c421-8273-41cb-9e11-8c96a2b30e10 1        <PublishedDataset 'cd27c421-8273-41cb-9e11-8c96a2b30e10':1 - EPA-InVitroDBV3.2-NVS_NR_hPPARa_summary_tcpl>        \n",
       "291d8013-662b-4b44-aa96-894be4473d59 1        <PublishedDataset '291d8013-662b-4b44-aa96-894be4473d59':1 - EPA-InVitroDBV3.2-ATG_PPRE_CIS_up_summary_tcpl>      \n",
       "f3ebdf70-976f-4b19-92a1-25b189fa13fa 1        <PublishedDataset 'f3ebdf70-976f-4b19-92a1-25b189fa13fa':1 - EPA-InVitroDBV3.2-ATG_PPARa_TRANS_up_summary_tcpl>   \n",
       "\n",
       "                                                   Endpoint name  \\\n",
       "id                                   version                       \n",
       "d651ef92-c12e-4eba-8979-8dc3f77fc7f3 1        ATG_PPARa_TRANS_dn   \n",
       "5a9cf864-520d-4ca0-b77f-8333aa8a3d5c 1        ATG_PPRE_CIS_dn      \n",
       "cd27c421-8273-41cb-9e11-8c96a2b30e10 1        NVS_NR_hPPARa        \n",
       "291d8013-662b-4b44-aa96-894be4473d59 1        ATG_PPRE_CIS_up      \n",
       "f3ebdf70-976f-4b19-92a1-25b189fa13fa 1        ATG_PPARa_TRANS_up   \n",
       "\n",
       "                                                                        Biological target  \\\n",
       "id                                   version                                                \n",
       "d651ef92-c12e-4eba-8979-8dc3f77fc7f3 1        regulation of transcription factor activity   \n",
       "5a9cf864-520d-4ca0-b77f-8333aa8a3d5c 1        regulation of transcription factor activity   \n",
       "cd27c421-8273-41cb-9e11-8c96a2b30e10 1        receptor binding                              \n",
       "291d8013-662b-4b44-aa96-894be4473d59 1        regulation of transcription factor activity   \n",
       "f3ebdf70-976f-4b19-92a1-25b189fa13fa 1        regulation of transcription factor activity   \n",
       "\n",
       "                                              Entrez gene ID for the molecular target  \\\n",
       "id                                   version                                            \n",
       "d651ef92-c12e-4eba-8979-8dc3f77fc7f3 1        5465                                      \n",
       "5a9cf864-520d-4ca0-b77f-8333aa8a3d5c 1        5465                                      \n",
       "cd27c421-8273-41cb-9e11-8c96a2b30e10 1        5465                                      \n",
       "291d8013-662b-4b44-aa96-894be4473d59 1        5465                                      \n",
       "f3ebdf70-976f-4b19-92a1-25b189fa13fa 1        5465                                      \n",
       "\n",
       "                                             Symbol  \\\n",
       "id                                   version          \n",
       "d651ef92-c12e-4eba-8979-8dc3f77fc7f3 1        PPARA   \n",
       "5a9cf864-520d-4ca0-b77f-8333aa8a3d5c 1        PPARA   \n",
       "cd27c421-8273-41cb-9e11-8c96a2b30e10 1        PPARA   \n",
       "291d8013-662b-4b44-aa96-894be4473d59 1        PPARA   \n",
       "f3ebdf70-976f-4b19-92a1-25b189fa13fa 1        PPARA   \n",
       "\n",
       "                                                                                     Gene name  \n",
       "id                                   version                                                    \n",
       "d651ef92-c12e-4eba-8979-8dc3f77fc7f3 1        peroxisome proliferator-activated receptor alpha  \n",
       "5a9cf864-520d-4ca0-b77f-8333aa8a3d5c 1        peroxisome proliferator-activated receptor alpha  \n",
       "cd27c421-8273-41cb-9e11-8c96a2b30e10 1        peroxisome proliferator-activated receptor alpha  \n",
       "291d8013-662b-4b44-aa96-894be4473d59 1        peroxisome proliferator-activated receptor alpha  \n",
       "f3ebdf70-976f-4b19-92a1-25b189fa13fa 1        peroxisome proliferator-activated receptor alpha  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns = [\n",
    "#    (\"Endpoint\", \"$.assay.component.endpoint\"),\n",
    "    (\"Endpoint name\", \"$.assay.component.endpoint.assay_component_endpoint_name.value\"),\n",
    "    (\"Biological target\", \"$.assay.component.endpoint.target.biological_process_target.value\"),\n",
    "    (\"Entrez gene ID for the molecular target\", \"$.assay.component.endpoint.target.intended.intended_target_gene.intended_target_entrez_gene_id.value\"),\n",
    "    (\"Symbol\", \"$.assay.component.endpoint.target.intended.intended_target_gene.intended_target_official_symbol.value\"),\n",
    "    (\"Gene name\", \"$.assay.component.endpoint.target.intended.intended_target_gene.intended_target_gene_name.value\")]\n",
    "\n",
    "cquery = None\n",
    "for gene in Genes:\n",
    "    if cquery is None:\n",
    "        cquery = Q.search_anywhere(\"EPA-InVitroDBV3.2\") & Q.search_anywhere(\"summary\") & Q.exact_search(Q.column('Entrez gene ID for the molecular target'), gene)\n",
    "    else:\n",
    "        cquery = cquery | Q.search_anywhere(\"EPA-InVitroDBV3.2\") & Q.search_anywhere(\"summary\") & Q.exact_search(Q.column('Entrez gene ID for the molecular target'), gene)\n",
    "\n",
    "ToxCast = api.get_published_datasets(limit=200, columns=columns, condition=cquery)\n",
    "ToxCast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Assay</th>\n",
       "      <th>DTXSID</th>\n",
       "      <th>Substance name</th>\n",
       "      <th>InChI key</th>\n",
       "      <th>CAS</th>\n",
       "      <th>IC50</th>\n",
       "      <th>Quality check</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3714</th>\n",
       "      <td>ATG_PPARa_TRANS_dn</td>\n",
       "      <td>DTXSID5020607</td>\n",
       "      <td>Di(2-ethylhexyl) phthalate</td>\n",
       "      <td>BJQHLKABXJIVAM-UHFFFAOYSA-N</td>\n",
       "      <td>117-81-7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3714</th>\n",
       "      <td>ATG_PPARa_TRANS_up</td>\n",
       "      <td>DTXSID5020607</td>\n",
       "      <td>Di(2-ethylhexyl) phthalate</td>\n",
       "      <td>BJQHLKABXJIVAM-UHFFFAOYSA-N</td>\n",
       "      <td>117-81-7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Borderline inactive]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3714</th>\n",
       "      <td>ATG_PPRE_CIS_dn</td>\n",
       "      <td>DTXSID5020607</td>\n",
       "      <td>Di(2-ethylhexyl) phthalate</td>\n",
       "      <td>BJQHLKABXJIVAM-UHFFFAOYSA-N</td>\n",
       "      <td>117-81-7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3714</th>\n",
       "      <td>ATG_PPRE_CIS_up</td>\n",
       "      <td>DTXSID5020607</td>\n",
       "      <td>Di(2-ethylhexyl) phthalate</td>\n",
       "      <td>BJQHLKABXJIVAM-UHFFFAOYSA-N</td>\n",
       "      <td>117-81-7</td>\n",
       "      <td>1.414849</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>NVS_NR_hPPARa</td>\n",
       "      <td>DTXSID5020607</td>\n",
       "      <td>Di(2-ethylhexyl) phthalate</td>\n",
       "      <td>BJQHLKABXJIVAM-UHFFFAOYSA-N</td>\n",
       "      <td>117-81-7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>919</th>\n",
       "      <td>ATG_PPARa_TRANS_dn</td>\n",
       "      <td>DTXSID0020652</td>\n",
       "      <td>Gemfibrozil</td>\n",
       "      <td>HEMJJKBWTPKOJG-UHFFFAOYSA-N</td>\n",
       "      <td>25812-30-0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>919</th>\n",
       "      <td>ATG_PPARa_TRANS_up</td>\n",
       "      <td>DTXSID0020652</td>\n",
       "      <td>Gemfibrozil</td>\n",
       "      <td>HEMJJKBWTPKOJG-UHFFFAOYSA-N</td>\n",
       "      <td>25812-30-0</td>\n",
       "      <td>1.558449</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>919</th>\n",
       "      <td>ATG_PPRE_CIS_dn</td>\n",
       "      <td>DTXSID0020652</td>\n",
       "      <td>Gemfibrozil</td>\n",
       "      <td>HEMJJKBWTPKOJG-UHFFFAOYSA-N</td>\n",
       "      <td>25812-30-0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>919</th>\n",
       "      <td>ATG_PPRE_CIS_up</td>\n",
       "      <td>DTXSID0020652</td>\n",
       "      <td>Gemfibrozil</td>\n",
       "      <td>HEMJJKBWTPKOJG-UHFFFAOYSA-N</td>\n",
       "      <td>25812-30-0</td>\n",
       "      <td>1.509815</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3245</th>\n",
       "      <td>ATG_PPARa_TRANS_dn</td>\n",
       "      <td>DTXSID3029869</td>\n",
       "      <td>Bezafibrate</td>\n",
       "      <td>IIBYAHWJQTYFKB-UHFFFAOYSA-N</td>\n",
       "      <td>41859-67-0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Noisy data]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3245</th>\n",
       "      <td>ATG_PPARa_TRANS_up</td>\n",
       "      <td>DTXSID3029869</td>\n",
       "      <td>Bezafibrate</td>\n",
       "      <td>IIBYAHWJQTYFKB-UHFFFAOYSA-N</td>\n",
       "      <td>41859-67-0</td>\n",
       "      <td>0.918967</td>\n",
       "      <td>[Hit-call potentially confounded by overfitting]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3245</th>\n",
       "      <td>ATG_PPRE_CIS_dn</td>\n",
       "      <td>DTXSID3029869</td>\n",
       "      <td>Bezafibrate</td>\n",
       "      <td>IIBYAHWJQTYFKB-UHFFFAOYSA-N</td>\n",
       "      <td>41859-67-0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Noisy data]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3245</th>\n",
       "      <td>ATG_PPRE_CIS_up</td>\n",
       "      <td>DTXSID3029869</td>\n",
       "      <td>Bezafibrate</td>\n",
       "      <td>IIBYAHWJQTYFKB-UHFFFAOYSA-N</td>\n",
       "      <td>41859-67-0</td>\n",
       "      <td>1.208669</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4050</th>\n",
       "      <td>ATG_PPARa_TRANS_dn</td>\n",
       "      <td>DTXSID3020336</td>\n",
       "      <td>Clofibrate</td>\n",
       "      <td>KNHUKKLJHYUCFP-UHFFFAOYSA-N</td>\n",
       "      <td>637-07-0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4050</th>\n",
       "      <td>ATG_PPARa_TRANS_up</td>\n",
       "      <td>DTXSID3020336</td>\n",
       "      <td>Clofibrate</td>\n",
       "      <td>KNHUKKLJHYUCFP-UHFFFAOYSA-N</td>\n",
       "      <td>637-07-0</td>\n",
       "      <td>1.653510</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4050</th>\n",
       "      <td>ATG_PPRE_CIS_dn</td>\n",
       "      <td>DTXSID3020336</td>\n",
       "      <td>Clofibrate</td>\n",
       "      <td>KNHUKKLJHYUCFP-UHFFFAOYSA-N</td>\n",
       "      <td>637-07-0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4050</th>\n",
       "      <td>ATG_PPRE_CIS_up</td>\n",
       "      <td>DTXSID3020336</td>\n",
       "      <td>Clofibrate</td>\n",
       "      <td>KNHUKKLJHYUCFP-UHFFFAOYSA-N</td>\n",
       "      <td>637-07-0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Borderline inactive]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>NVS_NR_hPPARa</td>\n",
       "      <td>DTXSID3020336</td>\n",
       "      <td>Clofibrate</td>\n",
       "      <td>KNHUKKLJHYUCFP-UHFFFAOYSA-N</td>\n",
       "      <td>637-07-0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2464</th>\n",
       "      <td>ATG_PPARa_TRANS_dn</td>\n",
       "      <td>DTXSID8020331</td>\n",
       "      <td>Ciprofibrate</td>\n",
       "      <td>KPSRODZRAIWAKH-UHFFFAOYSA-N</td>\n",
       "      <td>52214-84-3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Noisy data]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2464</th>\n",
       "      <td>ATG_PPARa_TRANS_up</td>\n",
       "      <td>DTXSID8020331</td>\n",
       "      <td>Ciprofibrate</td>\n",
       "      <td>KPSRODZRAIWAKH-UHFFFAOYSA-N</td>\n",
       "      <td>52214-84-3</td>\n",
       "      <td>0.009889</td>\n",
       "      <td>[Hit-call potentially confounded by overfitting]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2464</th>\n",
       "      <td>ATG_PPRE_CIS_dn</td>\n",
       "      <td>DTXSID8020331</td>\n",
       "      <td>Ciprofibrate</td>\n",
       "      <td>KPSRODZRAIWAKH-UHFFFAOYSA-N</td>\n",
       "      <td>52214-84-3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2464</th>\n",
       "      <td>ATG_PPRE_CIS_up</td>\n",
       "      <td>DTXSID8020331</td>\n",
       "      <td>Ciprofibrate</td>\n",
       "      <td>KPSRODZRAIWAKH-UHFFFAOYSA-N</td>\n",
       "      <td>52214-84-3</td>\n",
       "      <td>1.722385</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>ATG_PPARa_TRANS_dn</td>\n",
       "      <td>DTXSID4020290</td>\n",
       "      <td>Pirinixic acid</td>\n",
       "      <td>SZRPDCCEHVWOJX-UHFFFAOYSA-N</td>\n",
       "      <td>50892-23-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Noisy data]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>ATG_PPARa_TRANS_up</td>\n",
       "      <td>DTXSID4020290</td>\n",
       "      <td>Pirinixic acid</td>\n",
       "      <td>SZRPDCCEHVWOJX-UHFFFAOYSA-N</td>\n",
       "      <td>50892-23-4</td>\n",
       "      <td>0.745003</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>ATG_PPRE_CIS_dn</td>\n",
       "      <td>DTXSID4020290</td>\n",
       "      <td>Pirinixic acid</td>\n",
       "      <td>SZRPDCCEHVWOJX-UHFFFAOYSA-N</td>\n",
       "      <td>50892-23-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>ATG_PPRE_CIS_up</td>\n",
       "      <td>DTXSID4020290</td>\n",
       "      <td>Pirinixic acid</td>\n",
       "      <td>SZRPDCCEHVWOJX-UHFFFAOYSA-N</td>\n",
       "      <td>50892-23-4</td>\n",
       "      <td>1.307836</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>NVS_NR_hPPARa</td>\n",
       "      <td>DTXSID4020290</td>\n",
       "      <td>Pirinixic acid</td>\n",
       "      <td>SZRPDCCEHVWOJX-UHFFFAOYSA-N</td>\n",
       "      <td>50892-23-4</td>\n",
       "      <td>0.943310</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3855</th>\n",
       "      <td>ATG_PPARa_TRANS_dn</td>\n",
       "      <td>DTXSID2029874</td>\n",
       "      <td>Fenofibrate</td>\n",
       "      <td>YMTINGFKWWXKFG-UHFFFAOYSA-N</td>\n",
       "      <td>49562-28-9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Noisy data]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3855</th>\n",
       "      <td>ATG_PPARa_TRANS_up</td>\n",
       "      <td>DTXSID2029874</td>\n",
       "      <td>Fenofibrate</td>\n",
       "      <td>YMTINGFKWWXKFG-UHFFFAOYSA-N</td>\n",
       "      <td>49562-28-9</td>\n",
       "      <td>0.727661</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3855</th>\n",
       "      <td>ATG_PPRE_CIS_dn</td>\n",
       "      <td>DTXSID2029874</td>\n",
       "      <td>Fenofibrate</td>\n",
       "      <td>YMTINGFKWWXKFG-UHFFFAOYSA-N</td>\n",
       "      <td>49562-28-9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3855</th>\n",
       "      <td>ATG_PPRE_CIS_up</td>\n",
       "      <td>DTXSID2029874</td>\n",
       "      <td>Fenofibrate</td>\n",
       "      <td>YMTINGFKWWXKFG-UHFFFAOYSA-N</td>\n",
       "      <td>49562-28-9</td>\n",
       "      <td>1.191227</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>NVS_NR_hPPARa</td>\n",
       "      <td>DTXSID2029874</td>\n",
       "      <td>Fenofibrate</td>\n",
       "      <td>YMTINGFKWWXKFG-UHFFFAOYSA-N</td>\n",
       "      <td>49562-28-9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Assay         DTXSID              Substance name  \\\n",
       "3714  ATG_PPARa_TRANS_dn  DTXSID5020607  Di(2-ethylhexyl) phthalate   \n",
       "3714  ATG_PPARa_TRANS_up  DTXSID5020607  Di(2-ethylhexyl) phthalate   \n",
       "3714  ATG_PPRE_CIS_dn     DTXSID5020607  Di(2-ethylhexyl) phthalate   \n",
       "3714  ATG_PPRE_CIS_up     DTXSID5020607  Di(2-ethylhexyl) phthalate   \n",
       "24    NVS_NR_hPPARa       DTXSID5020607  Di(2-ethylhexyl) phthalate   \n",
       "919   ATG_PPARa_TRANS_dn  DTXSID0020652  Gemfibrozil                  \n",
       "919   ATG_PPARa_TRANS_up  DTXSID0020652  Gemfibrozil                  \n",
       "919   ATG_PPRE_CIS_dn     DTXSID0020652  Gemfibrozil                  \n",
       "919   ATG_PPRE_CIS_up     DTXSID0020652  Gemfibrozil                  \n",
       "3245  ATG_PPARa_TRANS_dn  DTXSID3029869  Bezafibrate                  \n",
       "3245  ATG_PPARa_TRANS_up  DTXSID3029869  Bezafibrate                  \n",
       "3245  ATG_PPRE_CIS_dn     DTXSID3029869  Bezafibrate                  \n",
       "3245  ATG_PPRE_CIS_up     DTXSID3029869  Bezafibrate                  \n",
       "4050  ATG_PPARa_TRANS_dn  DTXSID3020336  Clofibrate                   \n",
       "4050  ATG_PPARa_TRANS_up  DTXSID3020336  Clofibrate                   \n",
       "4050  ATG_PPRE_CIS_dn     DTXSID3020336  Clofibrate                   \n",
       "4050  ATG_PPRE_CIS_up     DTXSID3020336  Clofibrate                   \n",
       "353   NVS_NR_hPPARa       DTXSID3020336  Clofibrate                   \n",
       "2464  ATG_PPARa_TRANS_dn  DTXSID8020331  Ciprofibrate                 \n",
       "2464  ATG_PPARa_TRANS_up  DTXSID8020331  Ciprofibrate                 \n",
       "2464  ATG_PPRE_CIS_dn     DTXSID8020331  Ciprofibrate                 \n",
       "2464  ATG_PPRE_CIS_up     DTXSID8020331  Ciprofibrate                 \n",
       "132   ATG_PPARa_TRANS_dn  DTXSID4020290  Pirinixic acid               \n",
       "132   ATG_PPARa_TRANS_up  DTXSID4020290  Pirinixic acid               \n",
       "132   ATG_PPRE_CIS_dn     DTXSID4020290  Pirinixic acid               \n",
       "132   ATG_PPRE_CIS_up     DTXSID4020290  Pirinixic acid               \n",
       "409   NVS_NR_hPPARa       DTXSID4020290  Pirinixic acid               \n",
       "3855  ATG_PPARa_TRANS_dn  DTXSID2029874  Fenofibrate                  \n",
       "3855  ATG_PPARa_TRANS_up  DTXSID2029874  Fenofibrate                  \n",
       "3855  ATG_PPRE_CIS_dn     DTXSID2029874  Fenofibrate                  \n",
       "3855  ATG_PPRE_CIS_up     DTXSID2029874  Fenofibrate                  \n",
       "255   NVS_NR_hPPARa       DTXSID2029874  Fenofibrate                  \n",
       "\n",
       "                        InChI key         CAS      IC50  \\\n",
       "3714  BJQHLKABXJIVAM-UHFFFAOYSA-N  117-81-7   NaN         \n",
       "3714  BJQHLKABXJIVAM-UHFFFAOYSA-N  117-81-7   NaN         \n",
       "3714  BJQHLKABXJIVAM-UHFFFAOYSA-N  117-81-7   NaN         \n",
       "3714  BJQHLKABXJIVAM-UHFFFAOYSA-N  117-81-7    1.414849   \n",
       "24    BJQHLKABXJIVAM-UHFFFAOYSA-N  117-81-7   NaN         \n",
       "919   HEMJJKBWTPKOJG-UHFFFAOYSA-N  25812-30-0 NaN         \n",
       "919   HEMJJKBWTPKOJG-UHFFFAOYSA-N  25812-30-0  1.558449   \n",
       "919   HEMJJKBWTPKOJG-UHFFFAOYSA-N  25812-30-0 NaN         \n",
       "919   HEMJJKBWTPKOJG-UHFFFAOYSA-N  25812-30-0  1.509815   \n",
       "3245  IIBYAHWJQTYFKB-UHFFFAOYSA-N  41859-67-0 NaN         \n",
       "3245  IIBYAHWJQTYFKB-UHFFFAOYSA-N  41859-67-0  0.918967   \n",
       "3245  IIBYAHWJQTYFKB-UHFFFAOYSA-N  41859-67-0 NaN         \n",
       "3245  IIBYAHWJQTYFKB-UHFFFAOYSA-N  41859-67-0  1.208669   \n",
       "4050  KNHUKKLJHYUCFP-UHFFFAOYSA-N  637-07-0   NaN         \n",
       "4050  KNHUKKLJHYUCFP-UHFFFAOYSA-N  637-07-0    1.653510   \n",
       "4050  KNHUKKLJHYUCFP-UHFFFAOYSA-N  637-07-0   NaN         \n",
       "4050  KNHUKKLJHYUCFP-UHFFFAOYSA-N  637-07-0   NaN         \n",
       "353   KNHUKKLJHYUCFP-UHFFFAOYSA-N  637-07-0   NaN         \n",
       "2464  KPSRODZRAIWAKH-UHFFFAOYSA-N  52214-84-3 NaN         \n",
       "2464  KPSRODZRAIWAKH-UHFFFAOYSA-N  52214-84-3  0.009889   \n",
       "2464  KPSRODZRAIWAKH-UHFFFAOYSA-N  52214-84-3 NaN         \n",
       "2464  KPSRODZRAIWAKH-UHFFFAOYSA-N  52214-84-3  1.722385   \n",
       "132   SZRPDCCEHVWOJX-UHFFFAOYSA-N  50892-23-4 NaN         \n",
       "132   SZRPDCCEHVWOJX-UHFFFAOYSA-N  50892-23-4  0.745003   \n",
       "132   SZRPDCCEHVWOJX-UHFFFAOYSA-N  50892-23-4 NaN         \n",
       "132   SZRPDCCEHVWOJX-UHFFFAOYSA-N  50892-23-4  1.307836   \n",
       "409   SZRPDCCEHVWOJX-UHFFFAOYSA-N  50892-23-4  0.943310   \n",
       "3855  YMTINGFKWWXKFG-UHFFFAOYSA-N  49562-28-9 NaN         \n",
       "3855  YMTINGFKWWXKFG-UHFFFAOYSA-N  49562-28-9  0.727661   \n",
       "3855  YMTINGFKWWXKFG-UHFFFAOYSA-N  49562-28-9 NaN         \n",
       "3855  YMTINGFKWWXKFG-UHFFFAOYSA-N  49562-28-9  1.191227   \n",
       "255   YMTINGFKWWXKFG-UHFFFAOYSA-N  49562-28-9 NaN         \n",
       "\n",
       "                                         Quality check  \n",
       "3714  []                                                \n",
       "3714  [Borderline inactive]                             \n",
       "3714  []                                                \n",
       "3714  []                                                \n",
       "24    []                                                \n",
       "919   []                                                \n",
       "919   []                                                \n",
       "919   []                                                \n",
       "919   []                                                \n",
       "3245  [Noisy data]                                      \n",
       "3245  [Hit-call potentially confounded by overfitting]  \n",
       "3245  [Noisy data]                                      \n",
       "3245  []                                                \n",
       "4050  []                                                \n",
       "4050  []                                                \n",
       "4050  []                                                \n",
       "4050  [Borderline inactive]                             \n",
       "353   []                                                \n",
       "2464  [Noisy data]                                      \n",
       "2464  [Hit-call potentially confounded by overfitting]  \n",
       "2464  []                                                \n",
       "2464  []                                                \n",
       "132   [Noisy data]                                      \n",
       "132   []                                                \n",
       "132   []                                                \n",
       "132   []                                                \n",
       "409   []                                                \n",
       "3855  [Noisy data]                                      \n",
       "3855  []                                                \n",
       "3855  []                                                \n",
       "3855  []                                                \n",
       "255   []                                                "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cquery = None\n",
    "for compound in compoundstable['InChIKey'].values:\n",
    "    if cquery is None:\n",
    "        cquery = Q.fuzzy_search(Q.column('InChI key'), compound)\n",
    "    else:\n",
    "        cquery = cquery | Q.fuzzy_search(Q.column('InChI key'), compound)\n",
    "\n",
    "ToxCastData = pd.DataFrame()\n",
    "for index, row in ToxCast.iterrows():\n",
    "    tmpdata = row['dataset'].get_data(limit=None, condition=cquery)\n",
    "    tmpdata['Assay']=row['Endpoint name']\n",
    "    tmpdata = tmpdata[['Assay', 'DTXSID', 'Substance name', 'InChI key', 'CAS', 'IC50', 'Quality check']]\n",
    "    ToxCastData = pd.concat([ToxCastData, tmpdata])\n",
    "    \n",
    "ToxCastData.sort_values(by=['InChI key','Assay'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TG-GATES\n",
    "\n",
    "Based on the chemicals of interest, all transcriptomics datasets from TG-GATES are queried that are the result of exposure to those chemicals."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "      <th>Compound</th>\n",
       "      <th>InChI Key</th>\n",
       "      <th>CAS</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Organ</th>\n",
       "      <th>Study type</th>\n",
       "      <th>Dose</th>\n",
       "      <th>Dosing</th>\n",
       "      <th>Duration</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th>version</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7285e2a4-dd0d-4585-83b6-a258ed097042</th>\n",
       "      <th>1</th>\n",
       "      <td>&lt;PublishedDataset '7285e2a4-dd0d-4585-83b6-a258ed097042':1 - TG-GATES-gemfibrozil_Rat_Liver_in vivo_Repeat_15_day_high_FOLD_CHANGES&gt;</td>\n",
       "      <td>gemfibrozil</td>\n",
       "      <td>InChIKey=HEMJJKBWTPKOJG-UHFFFAOYSA-N</td>\n",
       "      <td>25812-30-0</td>\n",
       "      <td>Rat</td>\n",
       "      <td>Liver</td>\n",
       "      <td>in_vivo</td>\n",
       "      <td>high</td>\n",
       "      <td>Repeat</td>\n",
       "      <td>15day</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cb29047e-7d5f-45a2-94ab-510406abc89e</th>\n",
       "      <th>1</th>\n",
       "      <td>&lt;PublishedDataset 'cb29047e-7d5f-45a2-94ab-510406abc89e':1 - TG-GATES-gemfibrozil_Rat_Liver_in vivo_Repeat_29_day_low_FOLD_CHANGES&gt;</td>\n",
       "      <td>gemfibrozil</td>\n",
       "      <td>InChIKey=HEMJJKBWTPKOJG-UHFFFAOYSA-N</td>\n",
       "      <td>25812-30-0</td>\n",
       "      <td>Rat</td>\n",
       "      <td>Liver</td>\n",
       "      <td>in_vivo</td>\n",
       "      <td>low</td>\n",
       "      <td>Repeat</td>\n",
       "      <td>29day</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44307c2a-ee3d-4d01-9ef2-4551149eb528</th>\n",
       "      <th>1</th>\n",
       "      <td>&lt;PublishedDataset '44307c2a-ee3d-4d01-9ef2-4551149eb528':1 - TG-GATES-gemfibrozil_Rat_Liver_in vivo_Single_24_hr_high_FOLD_CHANGES&gt;</td>\n",
       "      <td>gemfibrozil</td>\n",
       "      <td>InChIKey=HEMJJKBWTPKOJG-UHFFFAOYSA-N</td>\n",
       "      <td>25812-30-0</td>\n",
       "      <td>Rat</td>\n",
       "      <td>Liver</td>\n",
       "      <td>in_vivo</td>\n",
       "      <td>high</td>\n",
       "      <td>Single</td>\n",
       "      <td>24hr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78400e01-a092-45be-9bbd-8365b924383f</th>\n",
       "      <th>1</th>\n",
       "      <td>&lt;PublishedDataset '78400e01-a092-45be-9bbd-8365b924383f':1 - TG-GATES-gemfibrozil_Rat_Liver_in vitro_2_hr_middle_FOLD_CHANGES&gt;</td>\n",
       "      <td>gemfibrozil</td>\n",
       "      <td>InChIKey=HEMJJKBWTPKOJG-UHFFFAOYSA-N</td>\n",
       "      <td>25812-30-0</td>\n",
       "      <td>Rat</td>\n",
       "      <td>Liver</td>\n",
       "      <td>in_vitro</td>\n",
       "      <td>middle</td>\n",
       "      <td>None</td>\n",
       "      <td>2hr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d9fbe763-8eb6-431c-97a9-6fc44307d27a</th>\n",
       "      <th>1</th>\n",
       "      <td>&lt;PublishedDataset 'd9fbe763-8eb6-431c-97a9-6fc44307d27a':1 - TG-GATES-gemfibrozil_Human_Liver_in vitro_2_hr_high_FOLD_CHANGES&gt;</td>\n",
       "      <td>gemfibrozil</td>\n",
       "      <td>InChIKey=HEMJJKBWTPKOJG-UHFFFAOYSA-N</td>\n",
       "      <td>25812-30-0</td>\n",
       "      <td>Human</td>\n",
       "      <td>Liver</td>\n",
       "      <td>in_vitro</td>\n",
       "      <td>high</td>\n",
       "      <td>None</td>\n",
       "      <td>2hr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>efeb9b4d-4315-4e56-b8a3-151bb4a79e77</th>\n",
       "      <th>1</th>\n",
       "      <td>&lt;PublishedDataset 'efeb9b4d-4315-4e56-b8a3-151bb4a79e77':1 - TG-GATES-fenofibrate_Rat_Liver_in vivo_Single_9_hr_low_FOLD_CHANGES&gt;</td>\n",
       "      <td>fenofibrate</td>\n",
       "      <td>InChIKey=YMTINGFKWWXKFG-UHFFFAOYSA-N</td>\n",
       "      <td>49562-28-9</td>\n",
       "      <td>Rat</td>\n",
       "      <td>Liver</td>\n",
       "      <td>in_vivo</td>\n",
       "      <td>low</td>\n",
       "      <td>Single</td>\n",
       "      <td>9hr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96834324-4a34-4314-bab4-a330cf4cefcd</th>\n",
       "      <th>1</th>\n",
       "      <td>&lt;PublishedDataset '96834324-4a34-4314-bab4-a330cf4cefcd':1 - TG-GATES-fenofibrate_Rat_Liver_in vivo_Single_24_hr_middle_FOLD_CHANGES&gt;</td>\n",
       "      <td>fenofibrate</td>\n",
       "      <td>InChIKey=YMTINGFKWWXKFG-UHFFFAOYSA-N</td>\n",
       "      <td>49562-28-9</td>\n",
       "      <td>Rat</td>\n",
       "      <td>Liver</td>\n",
       "      <td>in_vivo</td>\n",
       "      <td>middle</td>\n",
       "      <td>Single</td>\n",
       "      <td>24hr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361abe50-9559-4033-b614-fc913a26b7ae</th>\n",
       "      <th>1</th>\n",
       "      <td>&lt;PublishedDataset '361abe50-9559-4033-b614-fc913a26b7ae':1 - TG-GATES-fenofibrate_Rat_Liver_in vitro_2_hr_high_FOLD_CHANGES&gt;</td>\n",
       "      <td>fenofibrate</td>\n",
       "      <td>InChIKey=YMTINGFKWWXKFG-UHFFFAOYSA-N</td>\n",
       "      <td>49562-28-9</td>\n",
       "      <td>Rat</td>\n",
       "      <td>Liver</td>\n",
       "      <td>in_vitro</td>\n",
       "      <td>high</td>\n",
       "      <td>None</td>\n",
       "      <td>2hr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128264b2-6bed-4cfd-9896-90b7aec50a3f</th>\n",
       "      <th>1</th>\n",
       "      <td>&lt;PublishedDataset '128264b2-6bed-4cfd-9896-90b7aec50a3f':1 - TG-GATES-fenofibrate_Rat_Liver_in vivo_Single_6_hr_high_FOLD_CHANGES&gt;</td>\n",
       "      <td>fenofibrate</td>\n",
       "      <td>InChIKey=YMTINGFKWWXKFG-UHFFFAOYSA-N</td>\n",
       "      <td>49562-28-9</td>\n",
       "      <td>Rat</td>\n",
       "      <td>Liver</td>\n",
       "      <td>in_vivo</td>\n",
       "      <td>high</td>\n",
       "      <td>Single</td>\n",
       "      <td>6hr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ef2ba55a-7a08-41f3-878f-62465cc8c33d</th>\n",
       "      <th>1</th>\n",
       "      <td>&lt;PublishedDataset 'ef2ba55a-7a08-41f3-878f-62465cc8c33d':1 - TG-GATES-fenofibrate_Rat_Liver_in vitro_24_hr_high_FOLD_CHANGES&gt;</td>\n",
       "      <td>fenofibrate</td>\n",
       "      <td>InChIKey=YMTINGFKWWXKFG-UHFFFAOYSA-N</td>\n",
       "      <td>49562-28-9</td>\n",
       "      <td>Rat</td>\n",
       "      <td>Liver</td>\n",
       "      <td>in_vitro</td>\n",
       "      <td>high</td>\n",
       "      <td>None</td>\n",
       "      <td>24hr</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>187 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                            dataset  \\\n",
       "id                                   version                                                                                                                                          \n",
       "7285e2a4-dd0d-4585-83b6-a258ed097042 1        <PublishedDataset '7285e2a4-dd0d-4585-83b6-a258ed097042':1 - TG-GATES-gemfibrozil_Rat_Liver_in vivo_Repeat_15_day_high_FOLD_CHANGES>    \n",
       "cb29047e-7d5f-45a2-94ab-510406abc89e 1        <PublishedDataset 'cb29047e-7d5f-45a2-94ab-510406abc89e':1 - TG-GATES-gemfibrozil_Rat_Liver_in vivo_Repeat_29_day_low_FOLD_CHANGES>     \n",
       "44307c2a-ee3d-4d01-9ef2-4551149eb528 1        <PublishedDataset '44307c2a-ee3d-4d01-9ef2-4551149eb528':1 - TG-GATES-gemfibrozil_Rat_Liver_in vivo_Single_24_hr_high_FOLD_CHANGES>     \n",
       "78400e01-a092-45be-9bbd-8365b924383f 1        <PublishedDataset '78400e01-a092-45be-9bbd-8365b924383f':1 - TG-GATES-gemfibrozil_Rat_Liver_in vitro_2_hr_middle_FOLD_CHANGES>          \n",
       "d9fbe763-8eb6-431c-97a9-6fc44307d27a 1        <PublishedDataset 'd9fbe763-8eb6-431c-97a9-6fc44307d27a':1 - TG-GATES-gemfibrozil_Human_Liver_in vitro_2_hr_high_FOLD_CHANGES>          \n",
       "...                                                                                                                                                                      ...          \n",
       "efeb9b4d-4315-4e56-b8a3-151bb4a79e77 1        <PublishedDataset 'efeb9b4d-4315-4e56-b8a3-151bb4a79e77':1 - TG-GATES-fenofibrate_Rat_Liver_in vivo_Single_9_hr_low_FOLD_CHANGES>       \n",
       "96834324-4a34-4314-bab4-a330cf4cefcd 1        <PublishedDataset '96834324-4a34-4314-bab4-a330cf4cefcd':1 - TG-GATES-fenofibrate_Rat_Liver_in vivo_Single_24_hr_middle_FOLD_CHANGES>   \n",
       "361abe50-9559-4033-b614-fc913a26b7ae 1        <PublishedDataset '361abe50-9559-4033-b614-fc913a26b7ae':1 - TG-GATES-fenofibrate_Rat_Liver_in vitro_2_hr_high_FOLD_CHANGES>            \n",
       "128264b2-6bed-4cfd-9896-90b7aec50a3f 1        <PublishedDataset '128264b2-6bed-4cfd-9896-90b7aec50a3f':1 - TG-GATES-fenofibrate_Rat_Liver_in vivo_Single_6_hr_high_FOLD_CHANGES>      \n",
       "ef2ba55a-7a08-41f3-878f-62465cc8c33d 1        <PublishedDataset 'ef2ba55a-7a08-41f3-878f-62465cc8c33d':1 - TG-GATES-fenofibrate_Rat_Liver_in vitro_24_hr_high_FOLD_CHANGES>           \n",
       "\n",
       "                                                 Compound  \\\n",
       "id                                   version                \n",
       "7285e2a4-dd0d-4585-83b6-a258ed097042 1        gemfibrozil   \n",
       "cb29047e-7d5f-45a2-94ab-510406abc89e 1        gemfibrozil   \n",
       "44307c2a-ee3d-4d01-9ef2-4551149eb528 1        gemfibrozil   \n",
       "78400e01-a092-45be-9bbd-8365b924383f 1        gemfibrozil   \n",
       "d9fbe763-8eb6-431c-97a9-6fc44307d27a 1        gemfibrozil   \n",
       "...                                                   ...   \n",
       "efeb9b4d-4315-4e56-b8a3-151bb4a79e77 1        fenofibrate   \n",
       "96834324-4a34-4314-bab4-a330cf4cefcd 1        fenofibrate   \n",
       "361abe50-9559-4033-b614-fc913a26b7ae 1        fenofibrate   \n",
       "128264b2-6bed-4cfd-9896-90b7aec50a3f 1        fenofibrate   \n",
       "ef2ba55a-7a08-41f3-878f-62465cc8c33d 1        fenofibrate   \n",
       "\n",
       "                                                                         InChI Key  \\\n",
       "id                                   version                                         \n",
       "7285e2a4-dd0d-4585-83b6-a258ed097042 1        InChIKey=HEMJJKBWTPKOJG-UHFFFAOYSA-N   \n",
       "cb29047e-7d5f-45a2-94ab-510406abc89e 1        InChIKey=HEMJJKBWTPKOJG-UHFFFAOYSA-N   \n",
       "44307c2a-ee3d-4d01-9ef2-4551149eb528 1        InChIKey=HEMJJKBWTPKOJG-UHFFFAOYSA-N   \n",
       "78400e01-a092-45be-9bbd-8365b924383f 1        InChIKey=HEMJJKBWTPKOJG-UHFFFAOYSA-N   \n",
       "d9fbe763-8eb6-431c-97a9-6fc44307d27a 1        InChIKey=HEMJJKBWTPKOJG-UHFFFAOYSA-N   \n",
       "...                                                                            ...   \n",
       "efeb9b4d-4315-4e56-b8a3-151bb4a79e77 1        InChIKey=YMTINGFKWWXKFG-UHFFFAOYSA-N   \n",
       "96834324-4a34-4314-bab4-a330cf4cefcd 1        InChIKey=YMTINGFKWWXKFG-UHFFFAOYSA-N   \n",
       "361abe50-9559-4033-b614-fc913a26b7ae 1        InChIKey=YMTINGFKWWXKFG-UHFFFAOYSA-N   \n",
       "128264b2-6bed-4cfd-9896-90b7aec50a3f 1        InChIKey=YMTINGFKWWXKFG-UHFFFAOYSA-N   \n",
       "ef2ba55a-7a08-41f3-878f-62465cc8c33d 1        InChIKey=YMTINGFKWWXKFG-UHFFFAOYSA-N   \n",
       "\n",
       "                                                     CAS Organism  Organ  \\\n",
       "id                                   version                               \n",
       "7285e2a4-dd0d-4585-83b6-a258ed097042 1        25812-30-0  Rat      Liver   \n",
       "cb29047e-7d5f-45a2-94ab-510406abc89e 1        25812-30-0  Rat      Liver   \n",
       "44307c2a-ee3d-4d01-9ef2-4551149eb528 1        25812-30-0  Rat      Liver   \n",
       "78400e01-a092-45be-9bbd-8365b924383f 1        25812-30-0  Rat      Liver   \n",
       "d9fbe763-8eb6-431c-97a9-6fc44307d27a 1        25812-30-0  Human    Liver   \n",
       "...                                                  ...  ...        ...   \n",
       "efeb9b4d-4315-4e56-b8a3-151bb4a79e77 1        49562-28-9  Rat      Liver   \n",
       "96834324-4a34-4314-bab4-a330cf4cefcd 1        49562-28-9  Rat      Liver   \n",
       "361abe50-9559-4033-b614-fc913a26b7ae 1        49562-28-9  Rat      Liver   \n",
       "128264b2-6bed-4cfd-9896-90b7aec50a3f 1        49562-28-9  Rat      Liver   \n",
       "ef2ba55a-7a08-41f3-878f-62465cc8c33d 1        49562-28-9  Rat      Liver   \n",
       "\n",
       "                                             Study type    Dose  Dosing  \\\n",
       "id                                   version                              \n",
       "7285e2a4-dd0d-4585-83b6-a258ed097042 1        in_vivo    high    Repeat   \n",
       "cb29047e-7d5f-45a2-94ab-510406abc89e 1        in_vivo    low     Repeat   \n",
       "44307c2a-ee3d-4d01-9ef2-4551149eb528 1        in_vivo    high    Single   \n",
       "78400e01-a092-45be-9bbd-8365b924383f 1        in_vitro   middle  None     \n",
       "d9fbe763-8eb6-431c-97a9-6fc44307d27a 1        in_vitro   high    None     \n",
       "...                                                ...      ...   ...     \n",
       "efeb9b4d-4315-4e56-b8a3-151bb4a79e77 1        in_vivo    low     Single   \n",
       "96834324-4a34-4314-bab4-a330cf4cefcd 1        in_vivo    middle  Single   \n",
       "361abe50-9559-4033-b614-fc913a26b7ae 1        in_vitro   high    None     \n",
       "128264b2-6bed-4cfd-9896-90b7aec50a3f 1        in_vivo    high    Single   \n",
       "ef2ba55a-7a08-41f3-878f-62465cc8c33d 1        in_vitro   high    None     \n",
       "\n",
       "                                             Duration  \n",
       "id                                   version           \n",
       "7285e2a4-dd0d-4585-83b6-a258ed097042 1        15day    \n",
       "cb29047e-7d5f-45a2-94ab-510406abc89e 1        29day    \n",
       "44307c2a-ee3d-4d01-9ef2-4551149eb528 1        24hr     \n",
       "78400e01-a092-45be-9bbd-8365b924383f 1        2hr      \n",
       "d9fbe763-8eb6-431c-97a9-6fc44307d27a 1        2hr      \n",
       "...                                           ...      \n",
       "efeb9b4d-4315-4e56-b8a3-151bb4a79e77 1        9hr      \n",
       "96834324-4a34-4314-bab4-a330cf4cefcd 1        24hr     \n",
       "361abe50-9559-4033-b614-fc913a26b7ae 1        2hr      \n",
       "128264b2-6bed-4cfd-9896-90b7aec50a3f 1        6hr      \n",
       "ef2ba55a-7a08-41f3-878f-62465cc8c33d 1        24hr     \n",
       "\n",
       "[187 rows x 10 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns = [\n",
    "    (\"Compound\", \"$.Compound.Name\"),\n",
    "    (\"InChI Key\", \"$.Compound.\\\"InChI Key\\\"\"),\n",
    "    (\"CAS\", \"$.Compound.CAS\"),\n",
    "    (\"Organism\", \"$.Assay.Organism\"),\n",
    "    (\"Organ\", \"$.Assay.Organ\"),\n",
    "    (\"Study type\", \"$.Assay.\\\"Study type\\\"\"),\n",
    "    (\"Dose\", \"$.Assay.Exposure.Dose\"),\n",
    "    (\"Dosing\", \"$.Assay.Dosing\"),\n",
    "    (\"Duration\", \"$.Assay.Exposure.Duration\"),\n",
    "    (\"Duration unit\", \"$.Assay.Exposure.\\\"Duration unit\\\"\"),\n",
    "]\n",
    "\n",
    "cquery = None\n",
    "for compound in compoundstable['InChIKey'].values:\n",
    "    if cquery is None:\n",
    "        cquery = Q.fuzzy_search(Q.column('InChI Key'), compound)\n",
    "    else:\n",
    "        cquery = cquery | Q.fuzzy_search(Q.column('InChI Key'), compound)\n",
    "\n",
    "condition = Q.search_anywhere(\"TG-GATES\") & Q.search_anywhere(\"FOLD_CHANGES\") & (cquery)\n",
    "TGGATES = api.get_published_datasets(limit=api.get_raw_datasets(limit=0, columns=columns, condition=condition)['total'], columns=columns, condition=condition)\n",
    "TGGATES['Duration2'] = TGGATES['Duration'].map(str) + TGGATES['Duration unit']\n",
    "TGGATES['Duration'] = TGGATES['Duration2']\n",
    "TGGATES = TGGATES.drop(['Duration2','Duration unit'], axis=1)\n",
    "TGGATES = TGGATES.sort_values(by=['InChI Key'])\n",
    "TGGATES"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filtering the datasets\n",
    "\n",
    "In most cases, TG-GATES contains multiple datasets for the chemicals of interest, and should therefore be filtered. The next section allows the filtering of the datasets by the different columns from the TGGATES table. \n",
    "\n",
    "### Identify possible values per category\n",
    "First, all possible values for filters should be identified, summarizing the TGGATES table for the columns:\n",
    "- Compound\n",
    "- Organism\n",
    "- Organ\n",
    "- Study type\n",
    "- Dosing\n",
    "- Dose\n",
    "- Duration\n",
    "\n",
    "Second, a table will be generated with the possible compounds and their corresponding CAS IDs and InChI Keys. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data available for compounds: {'gemfibrozil', 'fenofibrate', 'WY-14643', 'clofibrate'}\n",
      "Data available for organisms: {'Rat', 'Human'}\n",
      "Data available for organs: {'Kidney', 'Liver'}\n",
      "Data available for study types: {'in_vivo', 'in_vitro'}\n",
      "Data available for dosings: {'Repeat', None, 'Single'}\n",
      "Data available for doses: {'low', 'middle', 'high'}\n",
      "Data available for durations: {'3hr', '24hr', '6hr', '15day', '8hr', '4day', '9hr', '8day', '2hr', '29day'}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CAS</th>\n",
       "      <th>InChI Key</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gemfibrozil</th>\n",
       "      <td>25812-30-0</td>\n",
       "      <td>InChIKey=HEMJJKBWTPKOJG-UHFFFAOYSA-N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>clofibrate</th>\n",
       "      <td>637-07-0</td>\n",
       "      <td>InChIKey=KNHUKKLJHYUCFP-UHFFFAOYSA-N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WY-14643</th>\n",
       "      <td>50892-23-4</td>\n",
       "      <td>InChIKey=SZRPDCCEHVWOJX-UHFFFAOYSA-N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fenofibrate</th>\n",
       "      <td>49562-28-9</td>\n",
       "      <td>InChIKey=YMTINGFKWWXKFG-UHFFFAOYSA-N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    CAS                             InChI Key\n",
       "gemfibrozil  25812-30-0  InChIKey=HEMJJKBWTPKOJG-UHFFFAOYSA-N\n",
       "clofibrate   637-07-0    InChIKey=KNHUKKLJHYUCFP-UHFFFAOYSA-N\n",
       "WY-14643     50892-23-4  InChIKey=SZRPDCCEHVWOJX-UHFFFAOYSA-N\n",
       "fenofibrate  49562-28-9  InChIKey=YMTINGFKWWXKFG-UHFFFAOYSA-N"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Options for categories\n",
    "compounds = set(TGGATES['Compound'].tolist())\n",
    "organisms = set(TGGATES['Organism'].tolist())\n",
    "organs = set(TGGATES['Organ'].tolist())\n",
    "studytypes = set(TGGATES['Study type'].tolist())\n",
    "dosings = set(TGGATES['Dosing'].tolist())\n",
    "doses = set(TGGATES['Dose'].tolist())\n",
    "durations = set(TGGATES['Duration'].tolist())\n",
    "\n",
    "print('Data available for compounds: '+str(compounds))\n",
    "print('Data available for organisms: '+str(organisms))\n",
    "print('Data available for organs: '+str(organs))\n",
    "print('Data available for study types: '+str(studytypes))\n",
    "print('Data available for dosings: '+str(dosings))\n",
    "print('Data available for doses: '+str(doses))\n",
    "print('Data available for durations: '+str(durations))\n",
    "\n",
    "#Table for compounds and corresponding identifiers\n",
    "df = pd.DataFrame()\n",
    "chemdict = {}\n",
    "\n",
    "for index, row in TGGATES.iterrows():\n",
    "    if not row['Compound'] in chemdict:\n",
    "        chemdict[row['Compound']] = {}\n",
    "        chemdict[row['Compound']]['CAS'] = row['CAS']\n",
    "        chemdict[row['Compound']]['InChI Key'] = row['InChI Key']\n",
    "\n",
    "df = pd.DataFrame.from_dict(chemdict, orient='index')\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter the datasets\n",
    "The values identified for each column can be used to filter the datasets. \n",
    "\n",
    "In order to do that, two lists should be filled:\n",
    "- A list of all categories involved in the filter, corresponding to the column header. This list is called ```list_cat```\n",
    "- A list of the filter values for all categories included in the category list. This list is called ```list_input```\n",
    "\n",
    "Note that the location of the filled values is important. The sequence of the categories, and their values should correspond between the two lists. For each category, one filter value can be filled in the filter list. One can enter as many filters as necessary, from filtering for only the organism variable to filtering for all possible variables captured in the TGGATES dataframe. The ```res_df``` table contains the dataset(s) that are used later for pathway analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Category</th>\n",
       "      <th>Input</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Compound</td>\n",
       "      <td>gemfibrozil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Organism</td>\n",
       "      <td>Rat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Organ</td>\n",
       "      <td>Liver</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Study type</td>\n",
       "      <td>in_vivo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Dosing</td>\n",
       "      <td>Repeat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Dose</td>\n",
       "      <td>high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Duration</td>\n",
       "      <td>8day</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Category        Input\n",
       "0  Compound    gemfibrozil\n",
       "1  Organism    Rat        \n",
       "2  Organ       Liver      \n",
       "3  Study type  in_vivo    \n",
       "4  Dosing      Repeat     \n",
       "5  Dose        high       \n",
       "6  Duration    8day       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "      <th>Compound</th>\n",
       "      <th>InChI Key</th>\n",
       "      <th>CAS</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Organ</th>\n",
       "      <th>Study type</th>\n",
       "      <th>Dose</th>\n",
       "      <th>Dosing</th>\n",
       "      <th>Duration</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th>version</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>e45c8706-010e-4597-b490-df138f57f06e</th>\n",
       "      <th>1</th>\n",
       "      <td>&lt;PublishedDataset 'e45c8706-010e-4597-b490-df138f57f06e':1 - TG-GATES-gemfibrozil_Rat_Liver_in vivo_Repeat_8_day_high_FOLD_CHANGES&gt;</td>\n",
       "      <td>gemfibrozil</td>\n",
       "      <td>InChIKey=HEMJJKBWTPKOJG-UHFFFAOYSA-N</td>\n",
       "      <td>25812-30-0</td>\n",
       "      <td>Rat</td>\n",
       "      <td>Liver</td>\n",
       "      <td>in_vivo</td>\n",
       "      <td>high</td>\n",
       "      <td>Repeat</td>\n",
       "      <td>8day</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                          dataset  \\\n",
       "id                                   version                                                                                                                                        \n",
       "e45c8706-010e-4597-b490-df138f57f06e 1        <PublishedDataset 'e45c8706-010e-4597-b490-df138f57f06e':1 - TG-GATES-gemfibrozil_Rat_Liver_in vivo_Repeat_8_day_high_FOLD_CHANGES>   \n",
       "\n",
       "                                                 Compound  \\\n",
       "id                                   version                \n",
       "e45c8706-010e-4597-b490-df138f57f06e 1        gemfibrozil   \n",
       "\n",
       "                                                                         InChI Key  \\\n",
       "id                                   version                                         \n",
       "e45c8706-010e-4597-b490-df138f57f06e 1        InChIKey=HEMJJKBWTPKOJG-UHFFFAOYSA-N   \n",
       "\n",
       "                                                     CAS Organism  Organ  \\\n",
       "id                                   version                               \n",
       "e45c8706-010e-4597-b490-df138f57f06e 1        25812-30-0  Rat      Liver   \n",
       "\n",
       "                                             Study type  Dose  Dosing Duration  \n",
       "id                                   version                                    \n",
       "e45c8706-010e-4597-b490-df138f57f06e 1        in_vivo    high  Repeat  8day     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "list_cat = [\"Compound\",\"Organism\",\"Organ\",\"Study type\",\"Dosing\",\"Dose\",\"Duration\"]\n",
    "list_input = [\"gemfibrozil\",\"Rat\",\"Liver\",\"in_vivo\",\"Repeat\",\"high\",\"8day\"]\n",
    "\n",
    "filterlist = pd.DataFrame(\n",
    "    {'Category': list_cat,\n",
    "     'Input': list_input\n",
    "    })\n",
    "display(filterlist)\n",
    "\n",
    "def tg_gates(df, list_cat, list_input):\n",
    "    comb = zip(list_cat, list_input)\n",
    "    sub_df = df\n",
    "    for tup in comb:\n",
    "        sub_df = sub_df[sub_df[tup[0]]==tup[1]]\n",
    "    return sub_df\n",
    "\n",
    "res_df = tg_gates(TGGATES, list_cat, list_input)\n",
    "display(res_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of differentially expressed genes: 50\n"
     ]
    }
   ],
   "source": [
    "deg = []\n",
    "for index, row in res_df.iterrows():\n",
    "    file = row['dataset'].get_data(limit=100000)\n",
    "    assay_deg = file[((file['logFC'] > 1) | (file['logFC'] < -1)) & (file['P.Value'] < 0.05) ]['ENTREZID'].unique().tolist()\n",
    "    if len(assay_deg) > 0:\n",
    "        for gene in assay_deg:\n",
    "            deg.append(gene)\n",
    "print('Number of differentially expressed genes: '+str(len(deg)))\n",
    "degforpwanalysis = set(deg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# WikiPathways RDF\n",
    "\n",
    "## Service description\n",
    "WikiPathways is a community-driven molecular pathway database, supporting wide-spread topics and supported by many databases and integrative resources. It contains semantic annotations in its pathways for genes, proteins, metabolites, and interactions using a variety of reference databases, and WikiPathways is used to analyze and integrate experimental omics datasets ([Slenter et al., 2017](https://doi.org/10.1093/nar/gkx1064)). Furthermore, human pathways from Reactome ([Fabregat et al., 2018](https://doi.org/10.1093/nar/gkx1132)), another molecular pathway database, are integrated with WikiPathways and are therefore part of the WikiPathways RDF ([Waagmeester et al., 2016](https://doi.org/10.1371/journal.pcbi.1004989)). On the OpenRiskNet e-infrastructure, the WikiPathways RDF, which includes the Reactome pathways, is exposed via a Virtuoso SPARQL endpoint.\n",
    "\n",
    "## Implementation\n",
    "The first section is to find all molecular pathways in WikiPathways that contain the genes of interest, by matching the results of a SPARQL query to the list of genes found with the AOP-DB RDF. The SPARQL query extracts all Entrez gene IDs from all pathways in WikiPathways. When overlap between those lists are found, the pathway ID is stored in a dataframe along with its title, and organism.\n",
    "\n",
    "The next section is for extracting all pathways for the species of interest, along with all genes present. These are used later for pathway analysis with the data extracted from TG-GATES through the EdelweissData explorer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pathway_title</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Entrez Gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WP1797</th>\n",
       "      <td>Circadian Clock</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[5465]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP3370</th>\n",
       "      <td>RORA activates gene expression</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[5465]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP299</th>\n",
       "      <td>Nuclear Receptors in Lipid Metabolism and Toxicity</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[5465]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP4721</th>\n",
       "      <td>Eicosanoid metabolism via Lipo Oxygenases (LOX)</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[5465]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP4447</th>\n",
       "      <td>SUMOylation of intracellular receptors</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[5465]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP4720</th>\n",
       "      <td>Eicosanoid metabolism via Cytochrome P450 Mono-Oxygenases (CYP) pathway</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[5465]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP2011</th>\n",
       "      <td>SREBF and miR33 in cholesterol and lipid homeostasis</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[5465]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP2882</th>\n",
       "      <td>Nuclear Receptors Meta-Pathway</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[5465]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP3594</th>\n",
       "      <td>Circadian rhythm related genes</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[5465]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP4396</th>\n",
       "      <td>Nonalcoholic fatty liver disease</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[5465]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP431</th>\n",
       "      <td>Nuclear receptors in lipid metabolism and toxicity</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>[19013]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP2084</th>\n",
       "      <td>SREBF and miR33 in cholesterol and lipid homeostasis</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>[19013]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP2316</th>\n",
       "      <td>PPAR signaling pathway</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>[19013]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP447</th>\n",
       "      <td>Adipogenesis genes</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>[19013]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP509</th>\n",
       "      <td>Nuclear Receptors</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>[19013]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP1099</th>\n",
       "      <td>Nuclear receptors in lipid metabolism and toxicity</td>\n",
       "      <td>Canis familiaris</td>\n",
       "      <td>[403654]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP1105</th>\n",
       "      <td>Adipogenesis</td>\n",
       "      <td>Canis familiaris</td>\n",
       "      <td>[403654]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP1184</th>\n",
       "      <td>Nuclear Receptors</td>\n",
       "      <td>Canis familiaris</td>\n",
       "      <td>[403654]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP1541</th>\n",
       "      <td>Energy Metabolism</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[5465]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP2706</th>\n",
       "      <td>Activation of gene expression by SREBF (SREBP)</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[5465]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP3942</th>\n",
       "      <td>PPAR signaling pathway</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[5465]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP2878</th>\n",
       "      <td>PPAR Alpha Pathway</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[5465]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP3355</th>\n",
       "      <td>BMAL1:CLOCK,NPAS2 activates circadian gene expression</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[5465]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP170</th>\n",
       "      <td>Nuclear Receptors</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[5465]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP2881</th>\n",
       "      <td>Estrogen Receptor Pathway</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[5465]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP2797</th>\n",
       "      <td>Regulation of lipid metabolism by Peroxisome proliferator-activated receptor alpha (PPARalpha)</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[5465]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP236</th>\n",
       "      <td>Adipogenesis</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[5465]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP3331</th>\n",
       "      <td>Mitochondrial biogenesis</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[5465]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP1822</th>\n",
       "      <td>Generic Transcription Pathway</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[5465]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP139</th>\n",
       "      <td>Nuclear receptors in lipid metabolism and toxicity</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>[25747]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP217</th>\n",
       "      <td>Nuclear Receptors</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>[25747]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP155</th>\n",
       "      <td>Adipogenesis</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>[25747]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP2751</th>\n",
       "      <td>Transcriptional regulation of white adipocyte differentiation</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[5465]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                         Pathway_title  \\\n",
       "WP1797  Circadian Clock                                                                                  \n",
       "WP3370  RORA activates gene expression                                                                   \n",
       "WP299   Nuclear Receptors in Lipid Metabolism and Toxicity                                               \n",
       "WP4721  Eicosanoid metabolism via Lipo Oxygenases (LOX)                                                  \n",
       "WP4447  SUMOylation of intracellular receptors                                                           \n",
       "WP4720  Eicosanoid metabolism via Cytochrome P450 Mono-Oxygenases (CYP) pathway                          \n",
       "WP2011  SREBF and miR33 in cholesterol and lipid homeostasis                                             \n",
       "WP2882  Nuclear Receptors Meta-Pathway                                                                   \n",
       "WP3594  Circadian rhythm related genes                                                                   \n",
       "WP4396  Nonalcoholic fatty liver disease                                                                 \n",
       "WP431   Nuclear receptors in lipid metabolism and toxicity                                               \n",
       "WP2084  SREBF and miR33 in cholesterol and lipid homeostasis                                             \n",
       "WP2316  PPAR signaling pathway                                                                           \n",
       "WP447   Adipogenesis genes                                                                               \n",
       "WP509   Nuclear Receptors                                                                                \n",
       "WP1099  Nuclear receptors in lipid metabolism and toxicity                                               \n",
       "WP1105  Adipogenesis                                                                                     \n",
       "WP1184  Nuclear Receptors                                                                                \n",
       "WP1541  Energy Metabolism                                                                                \n",
       "WP2706  Activation of gene expression by SREBF (SREBP)                                                   \n",
       "WP3942  PPAR signaling pathway                                                                           \n",
       "WP2878  PPAR Alpha Pathway                                                                               \n",
       "WP3355  BMAL1:CLOCK,NPAS2 activates circadian gene expression                                            \n",
       "WP170   Nuclear Receptors                                                                                \n",
       "WP2881  Estrogen Receptor Pathway                                                                        \n",
       "WP2797  Regulation of lipid metabolism by Peroxisome proliferator-activated receptor alpha (PPARalpha)   \n",
       "WP236   Adipogenesis                                                                                     \n",
       "WP3331  Mitochondrial biogenesis                                                                         \n",
       "WP1822  Generic Transcription Pathway                                                                    \n",
       "WP139   Nuclear receptors in lipid metabolism and toxicity                                               \n",
       "WP217   Nuclear Receptors                                                                                \n",
       "WP155   Adipogenesis                                                                                     \n",
       "WP2751  Transcriptional regulation of white adipocyte differentiation                                    \n",
       "\n",
       "                 Organism Entrez Gene  \n",
       "WP1797  Homo sapiens       [5465]      \n",
       "WP3370  Homo sapiens       [5465]      \n",
       "WP299   Homo sapiens       [5465]      \n",
       "WP4721  Homo sapiens       [5465]      \n",
       "WP4447  Homo sapiens       [5465]      \n",
       "WP4720  Homo sapiens       [5465]      \n",
       "WP2011  Homo sapiens       [5465]      \n",
       "WP2882  Homo sapiens       [5465]      \n",
       "WP3594  Homo sapiens       [5465]      \n",
       "WP4396  Homo sapiens       [5465]      \n",
       "WP431   Mus musculus       [19013]     \n",
       "WP2084  Mus musculus       [19013]     \n",
       "WP2316  Mus musculus       [19013]     \n",
       "WP447   Mus musculus       [19013]     \n",
       "WP509   Mus musculus       [19013]     \n",
       "WP1099  Canis familiaris   [403654]    \n",
       "WP1105  Canis familiaris   [403654]    \n",
       "WP1184  Canis familiaris   [403654]    \n",
       "WP1541  Homo sapiens       [5465]      \n",
       "WP2706  Homo sapiens       [5465]      \n",
       "WP3942  Homo sapiens       [5465]      \n",
       "WP2878  Homo sapiens       [5465]      \n",
       "WP3355  Homo sapiens       [5465]      \n",
       "WP170   Homo sapiens       [5465]      \n",
       "WP2881  Homo sapiens       [5465]      \n",
       "WP2797  Homo sapiens       [5465]      \n",
       "WP236   Homo sapiens       [5465]      \n",
       "WP3331  Homo sapiens       [5465]      \n",
       "WP1822  Homo sapiens       [5465]      \n",
       "WP139   Rattus norvegicus  [25747]     \n",
       "WP217   Rattus norvegicus  [25747]     \n",
       "WP155   Rattus norvegicus  [25747]     \n",
       "WP2751  Homo sapiens       [5465]      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sparqlquery = '''\n",
    "    SELECT DISTINCT (str(?wpid) as ?Pathway_ID) (str(?PW_Title) as ?Pathway_title) (fn:substring(?ncbiGeneId,33) as ?Entrez)  ?organism\n",
    "    WHERE {\n",
    "    ?gene a wp:GeneProduct; dcterms:identifier ?id; dcterms:isPartOf ?pathwayRes; wp:bdbEntrezGene ?ncbiGeneId.\n",
    "    ?pathwayRes a wp:Pathway; dcterms:identifier ?wpid; dc:title ?PW_Title; wp:organismName ?organism.}\n",
    "    '''\n",
    "wikipathwayssparql.setQuery(sparqlquery)\n",
    "wikipathwayssparql.setReturnFormat(JSON)  \n",
    "results = wikipathwayssparql.query().convert()\n",
    "\n",
    "def intersection(lst1, lst2): \n",
    "    lst3 = [value for value in lst1 if value in lst2] \n",
    "    return lst3 \n",
    "\n",
    "WikiPathwaysGenes = {}\n",
    "WikiPathwaysNames = {}\n",
    "WikiPathwaysOrganism = {}\n",
    "for result in results[\"results\"][\"bindings\"]:\n",
    "    WikiPathwaysGenes[result[\"Pathway_ID\"][\"value\"]] = set([])\n",
    "for result in results[\"results\"][\"bindings\"]:\n",
    "    WikiPathwaysGenes[result[\"Pathway_ID\"][\"value\"]].add(result[\"Entrez\"][\"value\"])\n",
    "for result in results[\"results\"][\"bindings\"]:\n",
    "    WikiPathwaysNames[result[\"Pathway_ID\"][\"value\"]] = result[\"Pathway_title\"][\"value\"]\n",
    "for result in results[\"results\"][\"bindings\"]:\n",
    "    WikiPathwaysOrganism[result[\"Pathway_ID\"][\"value\"]] = result[\"organism\"][\"value\"]\n",
    "\n",
    "\n",
    "for lst in WikiPathwaysGenes:\n",
    "    genematch = intersection(WikiPathwaysGenes[lst], Genes)\n",
    "    WikiPathwaysGenes[lst] = [WikiPathwaysNames[lst], WikiPathwaysOrganism[lst], genematch, len(genematch)]\n",
    "\n",
    "WPtable = pd.DataFrame.from_dict(WikiPathwaysGenes,orient='index',columns=['Pathway_title','Organism','Entrez Gene','nGenes'])\n",
    "WPtable = WPtable[WPtable.nGenes >= 1]\n",
    "WPtable = WPtable.drop(columns='nGenes')\n",
    "display(WPtable)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extracting patways for pathway analysis\n",
    "\n",
    "Prior to extracting pathways, the species of interest should be stored in a variable. Note that this should be the latin species name, and corresponding to the filtered dataset(s) from TG-GATES."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "OrganismFilter = 'Rattus norvegicus'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genes</th>\n",
       "      <th>nGenes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WP1286</th>\n",
       "      <td>{154516, 24861, 292155, 25279, 301264, 29738, 311257, 103690051, 302302, 24424, 494499, 81869, 361631, 316325, 574523, 396551, 293779, 116631, 292915, 25426, 65030, 246767, 65185, 24426, 192242, 286954, 305264, 304322, 25147, 24404, 307092, 113992, 299566, 308190, 114846, 108348061, 246245, 24422, 24192, 24902, 26760, 685402, 301517, 100910526, 310848, 24912, 500257, 361510, 295430, 353498, 81676, 25458, 500892, 363618, 310855, 171341, 154985, 29326, 297029, 286921, 58953, 116632, 246247, 24298, 500359, 116686, 100910462, 303218, 293451, 25315, 312495, 84406, 364476, 25428, 83783, 288108, 24296, 114700, 301595, 81924, 360268, 499302, 64352, 171445, 396527, 290623, 300850, 307838, 25086, 25355, 25427, 25146, 29725, 286989, 303669, 289197, 691394, 362228, 24297, 108348148, ...}</td>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP1290</th>\n",
       "      <td>{24508, 24577, 81525, 84359, 29884, 502004, 64547, 24185, 291100, 103694380, 500592, 309165, 100911660, 100360940, 309295, 117279, 25718, 60434, 292892, 364081, 362675, 492821, 25272, 296953, 78963, 687813, 78971, 25166, 83584, 25402, 24842, 316256, 64625, 290749, 114555, 314856, 25385, 312030, 64044, 140923, 24482, 309361, 25008, 64026, 24516, 116502, 314756, 156767, 25625, 103689977, 64314, 266610, 246775, 103690372, 287398, 246097, 362788, 116667, 64041, 294071, 246756, 24887, 246334, 25513, 84351, 316241, 24483, 25493, 60371, 114214, 64639, 63879, 24224, 293624, 311786, 58918, 81736, 306886}</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP1278</th>\n",
       "      <td>{84351, 309361, 313121, 116554, 116590, 309452, 29538, 311245, 103694380, 81736, 309165, 81780, 299331, 24481, 360640, 286908, 246756}</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP1279</th>\n",
       "      <td>{84581, 288588, 363481, 498109, 116590, 24185, 288533, 690966, 81649, 84577, 297893, 81504, 50658, 292778, 114486, 314322, 84582, 103695118, 81646, 58919, 100363500, 84578, 363287, 84580, 170922, 313845, 306516, 24400, 81674, 83503, 83828, 310784, 117526, 680149, 361580, 682902, 497672, 171150, 307485, 309361, 171104, 117017, 24790, 24516, 25636, 288651, 294236, 81673, 83805, 289561, 294018, 361365, 309224, 294693, 103690054, 291703, 308415, 84351, 363633, 24890, 24224, 170851, 84389, 293621, 367858, 373541, 314612, 314436, 54244, 81736, 170915, 363067, 266713, 303918}</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP1282</th>\n",
       "      <td>{368066, 25331, 690050, 300691, 689330, 171347, 24661, 24267, 81676}</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP547</th>\n",
       "      <td>{85253, 306761, 24233, 79224, 24946, 25439, 113959, 25619, 24548, 24903, 83580, 24366, 298566, 81750, 116669, 117512, 29251, 25048, 29243, 113936, 295703, 65051, 79126, 288001, 289055, 29333, 29436, 117517, 302470, 25268, 289395, 50692, 24232, 24648, 54249, 24153, 29687, 192262, 313421, 64036, 64459, 24234, 362634, 304917, 155012, 24231, 25584, 287527, 81509, 24441, 25407, 312705, 24617, 24237, 54243, 290757, 84007, 64023, 260320, 25692, 294257}</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP505</th>\n",
       "      <td>{25125, 84353, 317376, 50554, 25712, 161452, 59328, 59107, 50658, 25313, 314322, 59086, 85435, 25495, 114208, 24881, 60584, 300054, 311061, 156726, 445442, 29357, 25631, 25296, 24835, 103691556, 311071, 24516, 24373, 81516, 94188, 367264, 25671, 29200, 50689, 25353, 316742, 367218, 25639, 367100, 29591, 83837, 29610, 293621, 367858, 54244, 24617, 81736, 81810, 25124, 170915, 313477, 497010, 84598}</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP654</th>\n",
       "      <td>{300711, 25203, 114483, 54237, 25112, 58919, 492821, 78963, 24708, 25402, 24842, 497672, 311562, 116502, 399489, 680110, 114851, 300668, 25729, 24887, 114212, 25309, 362817, 24224, 58918, 298795, 100363502}</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP89</th>\n",
       "      <td>{24493, 24471, 64159, 84359, 362456, 117279, 140657, 299625, 313121, 24708, 78963, 116554, 83584, 25402, 29432, 25385, 64044, 24835, 140926, 64026, 24516, 266610, 25591, 103689977, 289014, 287398, 246097, 362491, 315994, 116667, 360748, 114214, 25309, 24224, 58918, 60374, 116685, 100363502, 29431}</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP81</th>\n",
       "      <td>{24233, 24232, 362119, 29687, 192262, 312705, 313421, 64036, 362634, 24237, 298566, 64023, 117512, 24231, 117517, 298288}</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>153 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     Genes  \\\n",
       "WP1286  {154516, 24861, 292155, 25279, 301264, 29738, 311257, 103690051, 302302, 24424, 494499, 81869, 361631, 316325, 574523, 396551, 293779, 116631, 292915, 25426, 65030, 246767, 65185, 24426, 192242, 286954, 305264, 304322, 25147, 24404, 307092, 113992, 299566, 308190, 114846, 108348061, 246245, 24422, 24192, 24902, 26760, 685402, 301517, 100910526, 310848, 24912, 500257, 361510, 295430, 353498, 81676, 25458, 500892, 363618, 310855, 171341, 154985, 29326, 297029, 286921, 58953, 116632, 246247, 24298, 500359, 116686, 100910462, 303218, 293451, 25315, 312495, 84406, 364476, 25428, 83783, 288108, 24296, 114700, 301595, 81924, 360268, 499302, 64352, 171445, 396527, 290623, 300850, 307838, 25086, 25355, 25427, 25146, 29725, 286989, 303669, 289197, 691394, 362228, 24297, 108348148, ...}   \n",
       "WP1290  {24508, 24577, 81525, 84359, 29884, 502004, 64547, 24185, 291100, 103694380, 500592, 309165, 100911660, 100360940, 309295, 117279, 25718, 60434, 292892, 364081, 362675, 492821, 25272, 296953, 78963, 687813, 78971, 25166, 83584, 25402, 24842, 316256, 64625, 290749, 114555, 314856, 25385, 312030, 64044, 140923, 24482, 309361, 25008, 64026, 24516, 116502, 314756, 156767, 25625, 103689977, 64314, 266610, 246775, 103690372, 287398, 246097, 362788, 116667, 64041, 294071, 246756, 24887, 246334, 25513, 84351, 316241, 24483, 25493, 60371, 114214, 64639, 63879, 24224, 293624, 311786, 58918, 81736, 306886}                                                                                                                                                                                           \n",
       "WP1278  {84351, 309361, 313121, 116554, 116590, 309452, 29538, 311245, 103694380, 81736, 309165, 81780, 299331, 24481, 360640, 286908, 246756}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \n",
       "WP1279  {84581, 288588, 363481, 498109, 116590, 24185, 288533, 690966, 81649, 84577, 297893, 81504, 50658, 292778, 114486, 314322, 84582, 103695118, 81646, 58919, 100363500, 84578, 363287, 84580, 170922, 313845, 306516, 24400, 81674, 83503, 83828, 310784, 117526, 680149, 361580, 682902, 497672, 171150, 307485, 309361, 171104, 117017, 24790, 24516, 25636, 288651, 294236, 81673, 83805, 289561, 294018, 361365, 309224, 294693, 103690054, 291703, 308415, 84351, 363633, 24890, 24224, 170851, 84389, 293621, 367858, 373541, 314612, 314436, 54244, 81736, 170915, 363067, 266713, 303918}                                                                                                                                                                                                                      \n",
       "WP1282  {368066, 25331, 690050, 300691, 689330, 171347, 24661, 24267, 81676}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 \n",
       "...                                                                      ...                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 \n",
       "WP547   {85253, 306761, 24233, 79224, 24946, 25439, 113959, 25619, 24548, 24903, 83580, 24366, 298566, 81750, 116669, 117512, 29251, 25048, 29243, 113936, 295703, 65051, 79126, 288001, 289055, 29333, 29436, 117517, 302470, 25268, 289395, 50692, 24232, 24648, 54249, 24153, 29687, 192262, 313421, 64036, 64459, 24234, 362634, 304917, 155012, 24231, 25584, 287527, 81509, 24441, 25407, 312705, 24617, 24237, 54243, 290757, 84007, 64023, 260320, 25692, 294257}                                                                                                                                                                                                                                                                                                                                                    \n",
       "WP505   {25125, 84353, 317376, 50554, 25712, 161452, 59328, 59107, 50658, 25313, 314322, 59086, 85435, 25495, 114208, 24881, 60584, 300054, 311061, 156726, 445442, 29357, 25631, 25296, 24835, 103691556, 311071, 24516, 24373, 81516, 94188, 367264, 25671, 29200, 50689, 25353, 316742, 367218, 25639, 367100, 29591, 83837, 29610, 293621, 367858, 54244, 24617, 81736, 81810, 25124, 170915, 313477, 497010, 84598}                                                                                                                                                                                                                                                                                                                                                                                                     \n",
       "WP654   {300711, 25203, 114483, 54237, 25112, 58919, 492821, 78963, 24708, 25402, 24842, 497672, 311562, 116502, 399489, 680110, 114851, 300668, 25729, 24887, 114212, 25309, 362817, 24224, 58918, 298795, 100363502}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       \n",
       "WP89    {24493, 24471, 64159, 84359, 362456, 117279, 140657, 299625, 313121, 24708, 78963, 116554, 83584, 25402, 29432, 25385, 64044, 24835, 140926, 64026, 24516, 266610, 25591, 103689977, 289014, 287398, 246097, 362491, 315994, 116667, 360748, 114214, 25309, 24224, 58918, 60374, 116685, 100363502, 29431}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \n",
       "WP81    {24233, 24232, 362119, 29687, 192262, 312705, 313421, 64036, 362634, 24237, 298566, 64023, 117512, 24231, 117517, 298288}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            \n",
       "\n",
       "        nGenes  \n",
       "WP1286  143     \n",
       "WP1290  78      \n",
       "WP1278  17      \n",
       "WP1279  74      \n",
       "WP1282  9       \n",
       "...    ..       \n",
       "WP547   61      \n",
       "WP505   54      \n",
       "WP654   27      \n",
       "WP89    39      \n",
       "WP81    16      \n",
       "\n",
       "[153 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sparqlquery = '''\n",
    "    SELECT DISTINCT (str(?wpid) as ?Pathway_ID) (str(?PW_Title) as ?Pathway_title) (fn:substring(?ncbiGeneId,33) as ?Entrez) \n",
    "    WHERE {\n",
    "    ?gene a wp:GeneProduct; dcterms:identifier ?id; dcterms:isPartOf ?pathwayRes; wp:bdbEntrezGene ?ncbiGeneId.\n",
    "    ?pathwayRes a wp:Pathway; dcterms:identifier ?wpid; dc:title ?PW_Title; wp:organismName \"'''+OrganismFilter+'''\"^^xsd:string.}\n",
    "    '''\n",
    "wikipathwayssparql.setQuery(sparqlquery)\n",
    "wikipathwayssparql.setReturnFormat(JSON)  \n",
    "results = wikipathwayssparql.query().convert()\n",
    "\n",
    "WikiPathwaysGenes = {}\n",
    "WikiPathwaysNames = {}\n",
    "for result in results[\"results\"][\"bindings\"]:\n",
    "    WikiPathwaysGenes[result[\"Pathway_ID\"][\"value\"]] = set([])\n",
    "for result in results[\"results\"][\"bindings\"]:\n",
    "    WikiPathwaysNames[result[\"Pathway_ID\"][\"value\"]] = result[\"Pathway_title\"][\"value\"]\n",
    "for result in results[\"results\"][\"bindings\"]:\n",
    "    WikiPathwaysGenes[result[\"Pathway_ID\"][\"value\"]].add(result[\"Entrez\"][\"value\"])\n",
    "\n",
    "for lst in WikiPathwaysGenes:\n",
    "    WikiPathwaysGenes[lst] = [WikiPathwaysGenes[lst],len(WikiPathwaysGenes[lst])]\n",
    "\n",
    "WPtable = pd.DataFrame.from_dict(WikiPathwaysGenes,orient='index',columns=['Genes','nGenes'])\n",
    "display(WPtable)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This section is the actual pathway analysis, using the differentially expressed genes from TG-GATES, and the molecular pathways from WikiPathways. With some basic statistics, a Z-score can be calculated for all pathways. A Z-score above 1.96 is considered significant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genes</th>\n",
       "      <th>nGenes</th>\n",
       "      <th>nSigGenes</th>\n",
       "      <th>percentSigGenes</th>\n",
       "      <th>Zscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WP1286</th>\n",
       "      <td>{154516, 24861, 292155, 25279, 301264, 29738, 311257, 103690051, 302302, 24424, 494499, 81869, 361631, 316325, 574523, 396551, 293779, 116631, 292915, 25426, 65030, 246767, 65185, 24426, 192242, 286954, 305264, 304322, 25147, 24404, 307092, 113992, 299566, 308190, 114846, 108348061, 246245, 24422, 24192, 24902, 26760, 685402, 301517, 100910526, 310848, 24912, 500257, 361510, 295430, 353498, 81676, 25458, 500892, 363618, 310855, 171341, 154985, 29326, 297029, 286921, 58953, 116632, 246247, 24298, 500359, 116686, 100910462, 303218, 293451, 25315, 312495, 84406, 364476, 25428, 83783, 288108, 24296, 114700, 301595, 81924, 360268, 499302, 64352, 171445, 396527, 290623, 300850, 307838, 25086, 25355, 25427, 25146, 29725, 286989, 303669, 289197, 691394, 362228, 24297, 108348148, ...}</td>\n",
       "      <td>143</td>\n",
       "      <td>4</td>\n",
       "      <td>2.797203</td>\n",
       "      <td>5.401926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP1307</th>\n",
       "      <td>{170670, 113976, 117243, 25330, 25413, 25288, 364975, 25062, 24849, 94340, 361676, 25014, 117035, 171155, 140547, 24539, 311569, 79223, 311849, 24158, 24538, 289481, 113965, 29367, 117543, 25757, 25287, 114024, 29740, 100911615, 25363, 298942, 64304, 25756, 50682}</td>\n",
       "      <td>35</td>\n",
       "      <td>3</td>\n",
       "      <td>8.571429</td>\n",
       "      <td>3.959527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP506</th>\n",
       "      <td>{170670, 291468, 113976, 25330, 117243, 25413, 25288, 364975, 25062, 94340, 361676, 25014, 117035, 171155, 24539, 100911186, 311569, 311849, 24158, 289481, 24538, 113965, 29367, 117543, 25757, 25287, 114024, 29740, 100911615, 298942, 25363, 64304, 25756, 50682}</td>\n",
       "      <td>34</td>\n",
       "      <td>3</td>\n",
       "      <td>8.823529</td>\n",
       "      <td>3.959527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP1297</th>\n",
       "      <td>{29563, 25271, 366791, 310903, 499985, 25703, 24172, 89826, 24705, 29184, 24539, 25086, 361801, 29646, 690953, 83574, 24706, 266603, 100145871, 293049, 25061, 154985, 25056, 155192, 24188, 313689, 116676, 362662, 685072, 246298, 114628, 292915, 432367, 24710, 64047, 25073, 353252, 312495, 100365047, 314264, 83783, 114106}</td>\n",
       "      <td>42</td>\n",
       "      <td>3</td>\n",
       "      <td>7.142857</td>\n",
       "      <td>3.959527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP419</th>\n",
       "      <td>{170670, 24158, 25363, 113976, 171142, 113965, 117035, 25413, 25288, 64304, 25757, 25287, 114024, 25541, 29740, 113956}</td>\n",
       "      <td>16</td>\n",
       "      <td>3</td>\n",
       "      <td>18.750000</td>\n",
       "      <td>3.959527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP139</th>\n",
       "      <td>{114700, 84356, 24646, 25270, 81924, 25279, 25664, 313210, 24307, 24891, 25303, 24705, 25086, 83569, 24706, 25682, 65035, 85264, 361523, 154985, 140668, 84385, 685072, 24297, 29277, 114628, 60351, 24873, 58852, 170913, 25428, 25747}</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>6.250000</td>\n",
       "      <td>2.517128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP145</th>\n",
       "      <td>{25384, 24538, 84497, 25675, 300438, 25292, 310900, 25428, 25728, 296371, 25081, 24539, 25073, 24207, 25080, 299858, 81782, 313210, 108348160, 24530}</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>2.517128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP2376</th>\n",
       "      <td>{116643, 140727, 85420, 100361457, 116590, 684969, 298947, 29224, 83472, 50658, 24451, 116554, 24424, 310392, 291796, 100158233, 81869, 497672, 24919, 29437, 366960, 29739, 24565, 85421, 24534, 287876, 50689, 497931, 304127, 140668, 24314, 29292, 24426, 361632, 295549, 83688, 25445, 25522, 25150, 58960, 24185, 114846, 108348061, 26760, 114495, 25365, 64188, 24842, 100360087, 679217, 25581, 171379, 25458, 24516, 24252, 171341, 114851, 286921, 85430, 361568, 29326, 297029, 117262, 57298, 116667, 25513, 301252, 301555, 117254, 116686, 170538, 24778, 170851, 25283, 293621, 54349, 54244, 25315, 681050, 84027, 25073, 170915, 83619, 117263, 305540, 24552, 24189, 25023, 29741, 25260, 497932, 24908, 64191, 300711, 100912585, 81649, 289623, 25352, 79255, 65052, ...}</td>\n",
       "      <td>167</td>\n",
       "      <td>2</td>\n",
       "      <td>1.197605</td>\n",
       "      <td>2.517128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP372</th>\n",
       "      <td>{170670, 113976, 117243, 25330, 25413, 25288, 364975, 25062, 24849, 94340, 361676, 25014, 117035, 171155, 140547, 24539, 311569, 79223, 311849, 24158, 24538, 289481, 113965, 29367, 25757, 25287, 114024, 100911615, 25363, 298942, 64304, 25756, 50682}</td>\n",
       "      <td>33</td>\n",
       "      <td>2</td>\n",
       "      <td>6.060606</td>\n",
       "      <td>2.517128</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     Genes  \\\n",
       "WP1286  {154516, 24861, 292155, 25279, 301264, 29738, 311257, 103690051, 302302, 24424, 494499, 81869, 361631, 316325, 574523, 396551, 293779, 116631, 292915, 25426, 65030, 246767, 65185, 24426, 192242, 286954, 305264, 304322, 25147, 24404, 307092, 113992, 299566, 308190, 114846, 108348061, 246245, 24422, 24192, 24902, 26760, 685402, 301517, 100910526, 310848, 24912, 500257, 361510, 295430, 353498, 81676, 25458, 500892, 363618, 310855, 171341, 154985, 29326, 297029, 286921, 58953, 116632, 246247, 24298, 500359, 116686, 100910462, 303218, 293451, 25315, 312495, 84406, 364476, 25428, 83783, 288108, 24296, 114700, 301595, 81924, 360268, 499302, 64352, 171445, 396527, 290623, 300850, 307838, 25086, 25355, 25427, 25146, 29725, 286989, 303669, 289197, 691394, 362228, 24297, 108348148, ...}   \n",
       "WP1307  {170670, 113976, 117243, 25330, 25413, 25288, 364975, 25062, 24849, 94340, 361676, 25014, 117035, 171155, 140547, 24539, 311569, 79223, 311849, 24158, 24538, 289481, 113965, 29367, 117543, 25757, 25287, 114024, 29740, 100911615, 25363, 298942, 64304, 25756, 50682}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             \n",
       "WP506   {170670, 291468, 113976, 25330, 117243, 25413, 25288, 364975, 25062, 94340, 361676, 25014, 117035, 171155, 24539, 100911186, 311569, 311849, 24158, 289481, 24538, 113965, 29367, 117543, 25757, 25287, 114024, 29740, 100911615, 298942, 25363, 64304, 25756, 50682}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                \n",
       "WP1297  {29563, 25271, 366791, 310903, 499985, 25703, 24172, 89826, 24705, 29184, 24539, 25086, 361801, 29646, 690953, 83574, 24706, 266603, 100145871, 293049, 25061, 154985, 25056, 155192, 24188, 313689, 116676, 362662, 685072, 246298, 114628, 292915, 432367, 24710, 64047, 25073, 353252, 312495, 100365047, 314264, 83783, 114106}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  \n",
       "WP419   {170670, 24158, 25363, 113976, 171142, 113965, 117035, 25413, 25288, 64304, 25757, 25287, 114024, 25541, 29740, 113956}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              \n",
       "WP139   {114700, 84356, 24646, 25270, 81924, 25279, 25664, 313210, 24307, 24891, 25303, 24705, 25086, 83569, 24706, 25682, 65035, 85264, 361523, 154985, 140668, 84385, 685072, 24297, 29277, 114628, 60351, 24873, 58852, 170913, 25428, 25747}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             \n",
       "WP145   {25384, 24538, 84497, 25675, 300438, 25292, 310900, 25428, 25728, 296371, 25081, 24539, 25073, 24207, 25080, 299858, 81782, 313210, 108348160, 24530}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                \n",
       "WP2376  {116643, 140727, 85420, 100361457, 116590, 684969, 298947, 29224, 83472, 50658, 24451, 116554, 24424, 310392, 291796, 100158233, 81869, 497672, 24919, 29437, 366960, 29739, 24565, 85421, 24534, 287876, 50689, 497931, 304127, 140668, 24314, 29292, 24426, 361632, 295549, 83688, 25445, 25522, 25150, 58960, 24185, 114846, 108348061, 26760, 114495, 25365, 64188, 24842, 100360087, 679217, 25581, 171379, 25458, 24516, 24252, 171341, 114851, 286921, 85430, 361568, 29326, 297029, 117262, 57298, 116667, 25513, 301252, 301555, 117254, 116686, 170538, 24778, 170851, 25283, 293621, 54349, 54244, 25315, 681050, 84027, 25073, 170915, 83619, 117263, 305540, 24552, 24189, 25023, 29741, 25260, 497932, 24908, 64191, 300711, 100912585, 81649, 289623, 25352, 79255, 65052, ...}                       \n",
       "WP372   {170670, 113976, 117243, 25330, 25413, 25288, 364975, 25062, 24849, 94340, 361676, 25014, 117035, 171155, 140547, 24539, 311569, 79223, 311849, 24158, 24538, 289481, 113965, 29367, 25757, 25287, 114024, 100911615, 25363, 298942, 64304, 25756, 50682}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            \n",
       "\n",
       "        nGenes  nSigGenes  percentSigGenes    Zscore  \n",
       "WP1286  143     4          2.797203         5.401926  \n",
       "WP1307  35      3          8.571429         3.959527  \n",
       "WP506   34      3          8.823529         3.959527  \n",
       "WP1297  42      3          7.142857         3.959527  \n",
       "WP419   16      3          18.750000        3.959527  \n",
       "WP139   32      2          6.250000         2.517128  \n",
       "WP145   20      2          10.000000        2.517128  \n",
       "WP2376  167     2          1.197605         2.517128  \n",
       "WP372   33      2          6.060606         2.517128  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ngenepres = []\n",
    "for index, row in WPtable.iterrows():\n",
    "    genepres = []\n",
    "    for gene in row['Genes']:\n",
    "        for sig in degforpwanalysis:\n",
    "            if gene == str(sig):\n",
    "                genepres.append(gene)\n",
    "    ngenepres.append(len(genepres))\n",
    "WPtable['nSigGenes'] = ngenepres\n",
    "WPtable['percentSigGenes'] = (WPtable['nSigGenes'] / WPtable['nGenes'])*100\n",
    "\n",
    "total = []\n",
    "for index, row in WPtable.iterrows():\n",
    "    total.append(row['nSigGenes'])\n",
    "\n",
    "StandardDeviation = statistics.stdev(total)\n",
    "ExpectedValue = (sum(total)/len(WPtable))\n",
    "\n",
    "WPtable['Zscore'] = (WPtable['nSigGenes'] - ExpectedValue)/StandardDeviation\n",
    "WPtable = WPtable.sort_values(by=['Zscore'], ascending=False)\n",
    "WPtable = WPtable[WPtable.Zscore >= 1.96]\n",
    "display(WPtable)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WP1286\tMetapathway biotransformation\n",
      "WP1307\tFatty Acid Beta Oxidation\n",
      "WP506\tFatty Acid Beta Oxidation\n",
      "WP1297\tRetinol metabolism\n",
      "WP419\tMitochondrial LC-Fatty Acid Beta-Oxidation\n",
      "WP139\tNuclear receptors in lipid metabolism and toxicity\n",
      "WP145\tStatin Pathway\n",
      "WP2376\tNuclear factor, erythroid-derived 2, like 2 signaling pathway\n",
      "WP372\tBeta Oxidation Meta Pathway\n",
      "\n",
      "Based on dataset(s): \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "      <th>Compound</th>\n",
       "      <th>InChI Key</th>\n",
       "      <th>CAS</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Organ</th>\n",
       "      <th>Study type</th>\n",
       "      <th>Dose</th>\n",
       "      <th>Dosing</th>\n",
       "      <th>Duration</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th>version</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>e45c8706-010e-4597-b490-df138f57f06e</th>\n",
       "      <th>1</th>\n",
       "      <td>&lt;PublishedDataset 'e45c8706-010e-4597-b490-df138f57f06e':1 - TG-GATES-gemfibrozil_Rat_Liver_in vivo_Repeat_8_day_high_FOLD_CHANGES&gt;</td>\n",
       "      <td>gemfibrozil</td>\n",
       "      <td>InChIKey=HEMJJKBWTPKOJG-UHFFFAOYSA-N</td>\n",
       "      <td>25812-30-0</td>\n",
       "      <td>Rat</td>\n",
       "      <td>Liver</td>\n",
       "      <td>in_vivo</td>\n",
       "      <td>high</td>\n",
       "      <td>Repeat</td>\n",
       "      <td>8day</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                          dataset  \\\n",
       "id                                   version                                                                                                                                        \n",
       "e45c8706-010e-4597-b490-df138f57f06e 1        <PublishedDataset 'e45c8706-010e-4597-b490-df138f57f06e':1 - TG-GATES-gemfibrozil_Rat_Liver_in vivo_Repeat_8_day_high_FOLD_CHANGES>   \n",
       "\n",
       "                                                 Compound  \\\n",
       "id                                   version                \n",
       "e45c8706-010e-4597-b490-df138f57f06e 1        gemfibrozil   \n",
       "\n",
       "                                                                         InChI Key  \\\n",
       "id                                   version                                         \n",
       "e45c8706-010e-4597-b490-df138f57f06e 1        InChIKey=HEMJJKBWTPKOJG-UHFFFAOYSA-N   \n",
       "\n",
       "                                                     CAS Organism  Organ  \\\n",
       "id                                   version                               \n",
       "e45c8706-010e-4597-b490-df138f57f06e 1        25812-30-0  Rat      Liver   \n",
       "\n",
       "                                             Study type  Dose  Dosing Duration  \n",
       "id                                   version                                    \n",
       "e45c8706-010e-4597-b490-df138f57f06e 1        in_vivo    high  Repeat  8day     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "SigPathways = list(WPtable.index)\n",
    "for WP in SigPathways:\n",
    "    print(WP + '\\t' +WikiPathwaysNames[WP])\n",
    "print('\\nBased on dataset(s): ')\n",
    "display(res_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPython 3.6.3\n",
      "IPython 7.9.0\n",
      "\n",
      "sys 3.6.3 (default, May 31 2019, 13:05:43) \n",
      "[GCC 4.8.5 20150623 (Red Hat 4.8.5-36)]\n",
      "pip 20.0.1\n",
      "SPARQLWrapper 1.8.5\n",
      "pandas 0.25.3\n",
      "json 2.0.9\n",
      "re 2.2.1\n",
      "requests 2.22.0\n",
      "warnings unknown\n",
      "pyvis 0.1.7.0\n",
      "matplotlib not installed\n",
      "numpy 1.18.1\n",
      "IPython 7.9.0\n",
      "urllib unknown\n",
      "seaborn not installed\n",
      "statistics unknown\n",
      "\n",
      "compiler   : GCC 4.8.5 20150623 (Red Hat 4.8.5-36)\n",
      "system     : Linux\n",
      "release    : 3.10.0-1062.1.2.el7.x86_64\n",
      "machine    : x86_64\n",
      "processor  : x86_64\n",
      "CPU cores  : 60\n",
      "interpreter: 64bit\n",
      " \n",
      "last updated: Thu Jan 23 2020 14:00:14 UTC\n"
     ]
    }
   ],
   "source": [
    "%load_ext watermark\n",
    "\n",
    "#python, ipython, packages, and machine characteristics\n",
    "%watermark -v -m -p sys,pip,SPARQLWrapper,pandas,json,re,requests,warnings,pyvis,matplotlib,numpy,IPython,urllib,seaborn,statistics\n",
    "\n",
    "#dte\n",
    "print(\" \")\n",
    "%watermark -u -n -t -z"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The underlying dataset for AOP-Wiki RDF: aop-wiki-xml-2019-07-01\n"
     ]
    }
   ],
   "source": [
    "sparqlquery = '''\n",
    "SELECT ?originaldata \n",
    "WHERE{\n",
    "?dataset a void:Dataset ; \n",
    "pav:createdWith ?originaldata .\n",
    "}'''\n",
    "aopwikisparql.setQuery(sparqlquery)\n",
    "aopwikisparql.setReturnFormat(JSON)  \n",
    "results = aopwikisparql.query().convert()\n",
    "\n",
    "Result = results[\"results\"][\"bindings\"][0]['originaldata'][\"value\"]\n",
    "\n",
    "print(\"The underlying dataset for AOP-Wiki RDF: \"+Result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The WikiPathways RDF used in this notebook: http://data.wikipathways.org/20191210/rdf/ created on 2019-12-09T23:28:23.591Z\n"
     ]
    }
   ],
   "source": [
    "sparqlquery = '''\n",
    "select distinct ?dataset ?date where {\n",
    "  ?dataset a void:Dataset ;\n",
    "  pav:createdOn ?date .\n",
    "}'''\n",
    "wikipathwayssparql.setQuery(sparqlquery)\n",
    "wikipathwayssparql.setReturnFormat(JSON)  \n",
    "results = wikipathwayssparql.query().convert()\n",
    "\n",
    "Result = results[\"results\"][\"bindings\"][0]['dataset'][\"value\"]\n",
    "Date = results[\"results\"][\"bindings\"][0]['date'][\"value\"]\n",
    "\n",
    "print(\"The WikiPathways RDF used in this notebook: \"+Result+ \" created on \"+ Date)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
